// @dynamic-labs/bitcoin/src/index.js
'use client'
import { assertPackageVersion } from '@dynamic-labs/assert-package-version';
import { version } from '../package.js';
export { BitcoinWalletConnector } from './connectors/BitcoinWalletConnector.js';
import '../_virtual/_tslib.js';
import 'sats-connect';
import 'bitcoinjs-lib';
import '@dynamic-labs/wallet-connector-core';
import '@dynamic-labs/utils';
export { BitcoinSatsConnectConnector } from './connectors/BitcoinSatsConnectConnector/BitcoinSatsConnectConnector.js';
import { MagicEdenConnector } from './connectors/MagicEdenConnector/MagicEdenConnector.js';
import { PhantomConnector } from './connectors/PhantomConnector/PhantomConnector.js';
import { OkxConnector } from './connectors/OkxConnector/OkxConnector.js';
import { OneKeyConnector } from './connectors/OneKeyConnector/OneKeyConnector.js';
import { OylConnector } from './connectors/OylConnector/OylConnector.js';
import { UnisatConnector } from './connectors/UnisatConnector/UnisatConnector.js';
export { UnisatConnector } from './connectors/UnisatConnector/UnisatConnector.js';
import { FallbackBitcoinConnector } from './connectors/FallbackBitcoinConnector/FallbackBitcoinConnector.js';
import { BinanceConnector } from './connectors/BinanceConnector/BinanceConnector.js';
import { BitgetConnector } from './connectors/BitgetConnector/BitgetConnector.js';
import { fetchBtcKitConnectors } from './utils/fetchBtcKitConnectors/fetchBtcKitConnectors.js';
export { fetchBtcKitConnectors } from './utils/fetchBtcKitConnectors/fetchBtcKitConnectors.js';
import { fetchSatsConnectConnectors } from './utils/fetchSatsConnectConnectors/fetchSatsConnectConnectors.js';
export { fetchSatsConnectConnectors } from './utils/fetchSatsConnectConnectors/fetchSatsConnectConnectors.js';
export { hasSatsConnectFeature } from './utils/hasSatsConnectFeature.js';
export { getMempoolApiUrl } from './utils/getMempoolApiUrl.js';
export { createPsbtOptions } from './utils/psbt/createSignPsbtOptions.js';
export { satoshisToBtc } from './utils/satoshisToBtc/satoshisToBtc.js';
export { BitcoinWallet } from './wallet/BitcoinWallet.js';
export { isBitcoinWallet } from './wallet/isBitcoinWallet/isBitcoinWallet.js';
export { BitcoinProviderHelper } from './bitcoinProviderHelper.js';
export { HTTP_STATUS_NOT_FOUND, HTTP_STATUS_TOO_MANY_REQUESTS } from './const.js';

assertPackageVersion('@dynamic-labs/bitcoin', version);
const BitcoinWalletConnectors = (props) => [
    ...fetchSatsConnectConnectors(props),
    ...fetchBtcKitConnectors(props),
    MagicEdenConnector,
    PhantomConnector,
    OkxConnector,
    OneKeyConnector,
    OylConnector,
    UnisatConnector,
    BitgetConnector,
    BinanceConnector,
    FallbackBitcoinConnector,
];

export { BitcoinWalletConnectors };
