// @dynamic-labs/bitcoin/src/connectors/UnisatConnector/UnisatConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../_virtual/_tslib.cjs');
var bitcoinjsLib = require('bitcoinjs-lib');
var walletConnectorCore = require('@dynamic-labs/wallet-connector-core');
var BitcoinWalletConnector = require('../BitcoinWalletConnector.cjs');
var createSignPsbtOptions = require('../../utils/psbt/createSignPsbtOptions.cjs');

class UnisatConnector extends BitcoinWalletConnector.BitcoinWalletConnector {
    constructor(opts) {
        var _a;
        super(Object.assign(Object.assign({}, opts), { overrideKey: (_a = opts.overrideKey) !== null && _a !== void 0 ? _a : 'unisat' }));
        this.name = 'UniSat';
        this.overrideKey = 'unisat';
    }
    // other connectors use the Unisat api and extend the UnisatConnector class
    // so we need to use a different name for each connector
    getConnectorName() {
        return `${this.name.replace(' ', '')}Connector`;
    }
    getAddress() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const didOpenInAppBrowser = this.openInAppBrowserIfRequired();
            if (didOpenInAppBrowser) {
                return;
            }
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const [address] = yield provider.requestAccounts();
            const publicKey = yield provider.getPublicKey();
            const bitcoinAddress = {
                address,
                publicKey,
            };
            yield this.setConnectedAccountWithAddresses({
                active: true,
                mainAddress: address,
                ordinalsAddress: bitcoinAddress,
            });
            return address;
        });
    }
    signMessage(messageToSign, options) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            if (options === null || options === void 0 ? void 0 : options.address) {
                walletConnectorCore.logger.debug(`[${this.getConnectorName()}] signMessage - Not possible to specify address to sign with`, {
                    address: options.address,
                    protocol: options.protocol,
                });
            }
            return provider.signMessage(messageToSign, options === null || options === void 0 ? void 0 : options.protocol);
        });
    }
    sendBitcoin(transaction) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            const connectedAddress = yield this.getAddress();
            if (!connectedAddress || !provider) {
                return;
            }
            return provider.sendBitcoin(transaction.recipientAddress, Number(transaction.amount));
        });
    }
    signPsbt(request) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const psbtFromBase64 = bitcoinjsLib.Psbt.fromBase64(request.unsignedPsbtBase64);
            const signedPsbtHex = yield provider.signPsbt(psbtFromBase64.toHex(), createSignPsbtOptions.createPsbtOptions(psbtFromBase64, request));
            return { signedPsbt: bitcoinjsLib.Psbt.fromHex(signedPsbtHex).toBase64() };
        });
    }
    signPsbts(requests) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const psbtHexs = [];
            const psbtOptions = [];
            for (const request of requests) {
                const psbtFromBase64 = bitcoinjsLib.Psbt.fromBase64(request.unsignedPsbtBase64);
                psbtHexs.push(psbtFromBase64.toHex());
                psbtOptions.push(createSignPsbtOptions.createPsbtOptions(psbtFromBase64, request));
            }
            const signedPsbtHexs = yield provider.signPsbts(psbtHexs, psbtOptions);
            return signedPsbtHexs.map((signedPsbtHex) => bitcoinjsLib.Psbt.fromHex(signedPsbtHex).toBase64());
        });
    }
    getConnectedAccounts() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return [];
            }
            return provider.getAccounts();
        });
    }
}

exports.UnisatConnector = UnisatConnector;
