// @dynamic-labs/algorand/src/AlgorandWalletConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../_virtual/_tslib.cjs');
var walletConnectorCore = require('@dynamic-labs/wallet-connector-core');
var utils = require('@dynamic-labs/utils');
var AlgorandWallet = require('./wallet/AlgorandWallet.cjs');

const HTTP_STATUS_TOO_MANY_REQUESTS = 429;
const HTTP_STATUS_NOT_FOUND = 404;
class AlgorandLocalStorageCache {
    constructor(key) {
        this.SESSION_STORAGE_KEY = `algorand_${key}_currentAddress`;
    }
    getCurrentAddress() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const addr = yield utils.getItemAsync(this.SESSION_STORAGE_KEY);
            if (!addr) {
                return undefined;
            }
            return addr;
        });
    }
    setCurrentAddress(addr) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return utils.setItemAsync(this.SESSION_STORAGE_KEY, addr);
        });
    }
    clearCurrentAddress() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return utils.removeItemAsync(this.SESSION_STORAGE_KEY);
        });
    }
}
class AlgorandWalletConnector extends walletConnectorCore.WalletConnectorBase {
    constructor(opts) {
        super(opts);
        this.ChainWallet = AlgorandWallet.AlgorandWallet;
        this.connectedChain = 'ALGO';
        this.supportedChains = ['ALGO'];
        this.cache = new AlgorandLocalStorageCache(opts.cacheKey);
    }
    getDeepLink() {
        return undefined;
    }
    getBalance(address) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const response = yield fetch(`https://indexer.algoexplorerapi.io/v2/accounts/${address}`);
            if (!response.ok) {
                // if the request fails due to rate limits, return cached value
                if (response.status === HTTP_STATUS_TOO_MANY_REQUESTS) {
                    return '0';
                }
                // new accounts not yet indexed will return a 404
                if (response.status === HTTP_STATUS_NOT_FOUND) {
                    return '0';
                }
                return undefined;
            }
            const accountInfo = yield response.json();
            if (!accountInfo.account) {
                return undefined;
            }
            const balance = accountInfo.account.amount.toString();
            return balance.toString();
        });
    }
    endSession() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            yield this.cache.clearCurrentAddress();
        });
    }
    getConnectedAccounts() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const currentAddress = yield this.cache.getCurrentAddress();
            return currentAddress ? [currentAddress] : [];
        });
    }
    getBlockExplorerUrlsForCurrentNetwork() {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            return ['https://explorer.bitquery.io/algorand'];
        });
    }
}

exports.AlgorandWalletConnector = AlgorandWalletConnector;
