// @dynamic-labs/bitcoin/src/connectors/OylConnector/OylConnector.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { Psbt } from 'bitcoinjs-lib';
import { BitcoinWalletConnector } from '../BitcoinWalletConnector.js';
import { createPsbtOptions } from '../../utils/psbt/createSignPsbtOptions.js';

class OylConnector extends BitcoinWalletConnector {
    constructor(opts) {
        super(Object.assign(Object.assign({}, opts), { overrideKey: 'oylwallet' }));
        this.name = 'Oyl Wallet';
        this.overrideKey = 'oylwallet';
    }
    getAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const addresses = yield provider.getAddresses();
            const ordinalsAddress = addresses.taproot;
            const paymentAddress = addresses.nativeSegwit;
            const mainAddress = (_a = ordinalsAddress === null || ordinalsAddress === void 0 ? void 0 : ordinalsAddress.address) !== null && _a !== void 0 ? _a : paymentAddress === null || paymentAddress === void 0 ? void 0 : paymentAddress.address;
            yield this.setConnectedAccountWithAddresses({
                active: true,
                mainAddress,
                ordinalsAddress,
                paymentAddress,
            });
            return mainAddress;
        });
    }
    signMessage(messageToSign, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { address, protocol } = options;
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const { signature } = yield provider.signMessage({
                address,
                message: messageToSign,
                // Oyl Wallet calls it just bip322 instead of bip322-simple
                protocol: protocol === 'bip322-simple' ? 'bip322' : protocol,
            });
            return signature;
        });
    }
    signPsbt(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const psbtFromBase64 = Psbt.fromBase64(request.unsignedPsbtBase64);
            const options = createPsbtOptions(psbtFromBase64, request);
            const { psbt: signedPsbtHex } = yield provider.signPsbt({
                broadcast: false,
                finalize: options.autoFinalized,
                psbt: psbtFromBase64.toHex(),
            });
            return { signedPsbt: Psbt.fromHex(signedPsbtHex).toBase64() };
        });
    }
}

export { OylConnector };
