// @dynamic-labs/bitcoin/src/connectors/OkxConnector/OkxConnector.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { Psbt } from 'bitcoinjs-lib';
import { logger } from '@dynamic-labs/wallet-connector-core';
import { BitcoinWalletConnector } from '../BitcoinWalletConnector.js';
import { createPsbtOptions } from '../../utils/psbt/createSignPsbtOptions.js';

class OkxConnector extends BitcoinWalletConnector {
    constructor(opts) {
        super(Object.assign(Object.assign({}, opts), { overrideKey: 'okxwalletbtc' }));
        this.name = 'OKX Wallet';
    }
    getAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            const didOpenInAppBrowser = this.openInAppBrowserIfRequired();
            if (didOpenInAppBrowser) {
                return;
            }
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const { address, publicKey } = yield provider.connect();
            const bitcoinAddress = {
                address,
                publicKey,
            };
            yield this.setConnectedAccountWithAddresses({
                active: true,
                mainAddress: address,
                ordinalsAddress: bitcoinAddress,
            });
            return address;
        });
    }
    signMessage(messageToSign, options) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            // default to bip322-simple
            const protocol = (_a = options === null || options === void 0 ? void 0 : options.protocol) !== null && _a !== void 0 ? _a : 'bip322-simple';
            if (options === null || options === void 0 ? void 0 : options.address) {
                logger.debug('[OkxConnector] signMessage - Not possible to specify address to sign with', {
                    address: options.address,
                    protocol,
                });
            }
            return provider.signMessage(messageToSign, protocol);
        });
    }
    sendBitcoin(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            const connectedAddress = yield this.getAddress();
            if (!connectedAddress || !provider) {
                return;
            }
            return provider.sendBitcoin(transaction.recipientAddress, transaction.amount.toString());
        });
    }
    signPsbt(request) {
        return __awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const psbtFromBase64 = Psbt.fromBase64(request.unsignedPsbtBase64);
            const signedPsbtHex = yield provider.signPsbt(psbtFromBase64.toHex(), createPsbtOptions(psbtFromBase64, request));
            return { signedPsbt: Psbt.fromHex(signedPsbtHex).toBase64() };
        });
    }
    signPsbts(requests) {
        return __awaiter(this, void 0, void 0, function* () {
            const provider = this.getProvider();
            if (!provider) {
                return;
            }
            const psbtHexs = [];
            const psbtOptions = [];
            for (const request of requests) {
                const psbt = Psbt.fromBase64(request.unsignedPsbtBase64);
                psbtHexs.push(psbt.toHex());
                psbtOptions.push(createPsbtOptions(psbt, request));
            }
            const signedPsbtHexs = yield provider.signPsbts(psbtHexs, psbtOptions);
            return signedPsbtHexs.map((signedPsbtHex) => Psbt.fromHex(signedPsbtHex).toBase64());
        });
    }
}

export { OkxConnector };
