// @dynamic-labs/sui-core/src/wallet/SuiWallet.js
'use client'
import { __awaiter } from '../../_virtual/_tslib.js';
import { Transaction } from '@mysten/sui/transactions';
import { Wallet } from '@dynamic-labs/wallet-connector-core';
import { DynamicError } from '@dynamic-labs/utils';
import { MIST_PER_SUI } from '../utils/constants/constants.js';

class SuiWallet extends Wallet {
    /**
     * Sends the native balance of the wallet to the given address.
     * @param amount - The amount of balance
     * @param toAddress - The address to send the balance to.
     * @returns The transaction as base64 encoded bcs.
     */
    sendBalance(_a) {
        return __awaiter(this, arguments, void 0, function* ({ amount, toAddress, }) {
            var _b;
            yield this._connector.connect();
            const account = yield this._connector.getWalletAccount();
            const chain = account === null || account === void 0 ? void 0 : account.chains[0];
            if (!account) {
                throw new DynamicError('Unable to retrieve Sui wallet account');
            }
            if (!chain) {
                throw new DynamicError('Unable to retrieve Sui active network');
            }
            const transactionFeature = (_b = this._connector.getFeatures()) === null || _b === void 0 ? void 0 : _b['sui:signAndExecuteTransaction'];
            if (!transactionFeature) {
                throw new DynamicError('Wallet does not support sui:signAndExecuteTransaction');
            }
            const transaction = new Transaction();
            const mistAmount = Number(amount) * MIST_PER_SUI;
            // Let the transaction calculate gas automatically
            // https://sdk.mystenlabs.com/typescript/transaction-building/gas#gas-price
            const [coin] = transaction.splitCoins(transaction.gas, [mistAmount]);
            transaction.transferObjects([coin], toAddress);
            const { bytes } = yield transactionFeature.signAndExecuteTransaction({
                account,
                chain,
                transaction,
            });
            return bytes;
        });
    }
    /**
     * Returns the Sui Client object initialized for the wallet's current network.
     * @returns The [SuiClient] object.
     */
    getSuiClient() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getSuiClient();
        });
    }
    /**
     * @deprecated Use `getSuiClient` instead
     */
    getWalletClient() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getSuiClient();
        });
    }
    /**
     * Returns the wallet's current active account.
     * @returns a readonly [WalletAccount] object.
     */
    getWalletAccount() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._connector.getWalletAccount();
        });
    }
    /**
     * Returns the wallet's current active network.
     * @returns a string representing the active network, i.e. 'sui:devnet'.
     */
    getActiveNetwork() {
        return __awaiter(this, void 0, void 0, function* () {
            const account = yield this._connector.getWalletAccount();
            return account === null || account === void 0 ? void 0 : account.chains[0];
        });
    }
    /**
     * Signs a Sui [Transaction].
     * @param tx - The [Transaction] to sign.
     * @returns The signature of the signed transaction.
     */
    signTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            var _a;
            yield this._connector.connect();
            const account = yield this._connector.getWalletAccount();
            const chain = account === null || account === void 0 ? void 0 : account.chains[0];
            if (!account) {
                throw new DynamicError('Unable to retrieve Sui wallet account');
            }
            if (!chain) {
                throw new DynamicError('Unable to retrieve Sui active network');
            }
            const transactionFeature = (_a = this._connector.getFeatures()) === null || _a === void 0 ? void 0 : _a['sui:signTransaction'];
            if (!transactionFeature) {
                throw new DynamicError('Wallet does not support sui:signTransaction');
            }
            const signedTransaction = yield transactionFeature.signTransaction({
                account,
                chain,
                transaction,
            });
            return signedTransaction;
        });
    }
}

export { SuiWallet };
