// @dynamic-labs/solana/src/index.js
'use client'
import { assertPackageVersion } from '@dynamic-labs/assert-package-version';
import { version } from '../package.js';
export { SolanaWalletConnector, isSolanaWallet } from '@dynamic-labs/solana-core';
export { BackpackSol as BackpackConnector } from './injected/BackpackSol/BackpackSol.js';
export { InjectedWalletBase as SolanaInjectedConnector } from './injected/InjectedWalletBase/InjectedWalletBase.js';
export { SolanaWalletConnectors } from './SolanaWalletConnectors.js';
export { SolanaWalletConnectorsWithConfig } from './SolanaWalletConnectorsWithConfig/SolanaWalletConnectorsWithConfig.js';
export { isBackpackSolanaSigner } from './utils/isBackpackSolanaSigner.js';
export { isSignedMessage } from './utils/isSignedMessage.js';
import '@dynamic-labs/wallet-book';
import '@dynamic-labs/wallet-connector-core';
import '@dynamic-labs/utils';
import '../_virtual/_tslib.js';
import './utils/logger.js';
export { createSolanaSignerFromWalletStandard } from './injected/walletStandard/createSolanaSignerFromWalletStandard/createSolanaSignerFromWalletStandard.js';
export { getWalletStandardWallets } from './injected/walletStandard/getWalletStandardWallets/getWalletStandardWallets.js';
export { hasAllWalletStandardRequiredFeatures } from './injected/walletStandard/hasAllWalletStandardRequiredFeatures/hasAllWalletStandardRequiredFeatures.js';

/* eslint-disable @typescript-eslint/no-unused-vars */
assertPackageVersion('@dynamic-labs/solana', version);
