// @dynamic-labs/cosmos/src/connectors/CompassWalletConnector.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var CosmosWalletConnector = require('./CosmosWalletConnector.cjs');

class CompassWalletConnector extends CosmosWalletConnector.CosmosWalletConnector {
    constructor() {
        super(...arguments);
        this.name = 'compasscosmos';
        this.overrideKey = 'compasscosmos';
    }
    getWallet() {
        if (typeof window === 'undefined') {
            throw new Error('Compass is not available');
        }
        return window.compass;
    }
    getChainId() {
        const hasSei = this.chainIdMapping.find((chain) => chain.cosmosNetworkId === 'pacific-1');
        // Compass wallet only supports Sei
        if (!hasSei) {
            throw new Error('Sei is not available on this site');
        }
        else {
            return 'pacific-1';
        }
    }
    setupEventListeners() {
        window.addEventListener('leap_keystorechange', this._handleAccountChange);
    }
    teardownEventListeners() {
        window.removeEventListener('leap_keystorechange', this._handleAccountChange);
        this.setChainId(null);
    }
}

exports.CompassWalletConnector = CompassWalletConnector;
