// @dynamic-labs/starknet/src/wallets/argent/webwallet/webwallet.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../../../../_virtual/_tslib.cjs');
var starknetkit = require('starknetkit');
var webwallet = require('starknetkit/webwallet');
var walletConnectorCore = require('@dynamic-labs/wallet-connector-core');
var base = require('../base.cjs');

class WebWallet extends base.ArgentBase {
    constructor(opts) {
        super('Argent X - Web (Deprecated)', 'argentWebWallet', 'argentxweb', opts);
        this.overrideKey = 'argentxweb';
        this.canConnectViaCustodialService = true;
        this.disconnectHandler = () => { };
    }
    isInstalledOnBrowser() {
        return false;
    }
    prompt(options) {
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            const { connectorData, wallet } = yield starknetkit.connect({
                connectors: [new webwallet.WebWalletConnector({ url: 'https://web.argent.xyz' })],
                modalMode: options.silently ? 'neverAsk' : 'canAsk',
            });
            return {
                data: {
                    account: connectorData === null || connectorData === void 0 ? void 0 : connectorData.account,
                    chainId: connectorData === null || connectorData === void 0 ? void 0 : connectorData.chainId,
                },
                wallet: wallet !== null && wallet !== void 0 ? wallet : undefined,
            };
        });
    }
    setupEventListeners() {
        super.setupEventListeners();
        this.disconnectHandler = () => {
            walletConnectorCore.logger.debug('[WebWallet] on logout');
            this.endSession();
            this.emit('disconnect');
        };
        document.addEventListener(webwallet.WEBWALLET_LOGOUT_EVENT, this.disconnectHandler);
    }
    teardownEventListeners() {
        const _super = Object.create(null, {
            teardownEventListeners: { get: () => super.teardownEventListeners }
        });
        return _tslib.__awaiter(this, void 0, void 0, function* () {
            _super.teardownEventListeners.call(this);
            if (this.disconnectHandler) {
                document.removeEventListener(webwallet.WEBWALLET_LOGOUT_EVENT, this.disconnectHandler);
            }
        });
    }
}

exports.WebWallet = WebWallet;
