// starknetkit/dist/starknetkit.js
var Ee = Object.defineProperty;
var Pe = (t, e, r) => e in t ? Ee(t, e, { enumerable: !0, configurable: !0, writable: !0, value: r }) : t[e] = r;
var J = (t, e, r) => (Pe(t, typeof e != "symbol" ? e + "" : e, r), r);
import { m as Z } from "./core-d21d2e96.js";
import { r as pe, s as ze } from "./lastConnected-e9351912.js";
import "starknet";
import { W as Le, D as je } from "./index-8edbd361.js";
import { A as We } from "./index-d4f30f2e.js";
import { BraavosMobileBaseConnector as Te } from "./braavosMobile.js";
import { I as A, A as ee } from "./index-9bb48f8a.js";
import "./getStarknetChainId-7c4b3163.js";
const Re = () => {
  const t = navigator.userAgent.toLowerCase(), e = /android|webos|iphone|ipad|ipod|blackberry|windows phone/.test(t), r = "ontouchstart" in window || navigator.maxTouchPoints > 0, n = window.innerWidth <= 768;
  return e && (r || n);
}, Ie = ({
  argentMobileOptions: t,
  webWalletUrl: e
}) => {
  const r = typeof window < "u" ? /^((?!chrome|android).)*safari/i.test(navigator.userAgent) : !1, n = [];
  return r || (n.push(
    new A({ options: { id: "argentX" } })
  ), n.push(
    new A({ options: { id: "braavos" } })
  ), A.isWalletInjected("metamask") && n.push(
    new A({ options: { id: "metamask" } })
  ), A.isWalletInjected("fordefi") && n.push(
    new A({ options: { id: "fordefi" } })
  ), A.isWalletInjected("keplr") && n.push(
    new A({ options: { id: "keplr" } })
  )), n.push(new We(t)), Re() && n.push(new Te()), n.push(new Le({ url: e })), n;
}, Ve = {
  "Amazon Silk": "amazon_silk",
  "Android Browser": "android",
  Bada: "bada",
  BlackBerry: "blackberry",
  Chrome: "chrome",
  Chromium: "chromium",
  Electron: "electron",
  Epiphany: "epiphany",
  Firefox: "firefox",
  Focus: "focus",
  Generic: "generic",
  "Google Search": "google_search",
  Googlebot: "googlebot",
  "Internet Explorer": "ie",
  "K-Meleon": "k_meleon",
  Maxthon: "maxthon",
  "Microsoft Edge": "edge",
  "MZ Browser": "mz",
  "NAVER Whale Browser": "naver",
  Opera: "opera",
  "Opera Coast": "opera_coast",
  PhantomJS: "phantomjs",
  Puffin: "puffin",
  QupZilla: "qupzilla",
  QQ: "qq",
  QQLite: "qqlite",
  Safari: "safari",
  Sailfish: "sailfish",
  "Samsung Internet for Android": "samsung_internet",
  SeaMonkey: "seamonkey",
  Sleipnir: "sleipnir",
  Swing: "swing",
  Tizen: "tizen",
  "UC Browser": "uc",
  Vivaldi: "vivaldi",
  "WebOS Browser": "webos",
  WeChat: "wechat",
  "Yandex Browser": "yandex",
  Roku: "roku"
}, be = {
  amazon_silk: "Amazon Silk",
  android: "Android Browser",
  bada: "Bada",
  blackberry: "BlackBerry",
  chrome: "Chrome",
  chromium: "Chromium",
  electron: "Electron",
  epiphany: "Epiphany",
  firefox: "Firefox",
  focus: "Focus",
  generic: "Generic",
  googlebot: "Googlebot",
  google_search: "Google Search",
  ie: "Internet Explorer",
  k_meleon: "K-Meleon",
  maxthon: "Maxthon",
  edge: "Microsoft Edge",
  mz: "MZ Browser",
  naver: "NAVER Whale Browser",
  opera: "Opera",
  opera_coast: "Opera Coast",
  phantomjs: "PhantomJS",
  puffin: "Puffin",
  qupzilla: "QupZilla",
  qq: "QQ Browser",
  qqlite: "QQ Browser Lite",
  safari: "Safari",
  sailfish: "Sailfish",
  samsung_internet: "Samsung Internet for Android",
  seamonkey: "SeaMonkey",
  sleipnir: "Sleipnir",
  swing: "Swing",
  tizen: "Tizen",
  uc: "UC Browser",
  vivaldi: "Vivaldi",
  webos: "WebOS Browser",
  wechat: "WeChat",
  yandex: "Yandex Browser"
}, M = {
  tablet: "tablet",
  mobile: "mobile",
  desktop: "desktop",
  tv: "tv"
}, F = {
  WindowsPhone: "Windows Phone",
  Windows: "Windows",
  MacOS: "macOS",
  iOS: "iOS",
  Android: "Android",
  WebOS: "WebOS",
  BlackBerry: "BlackBerry",
  Bada: "Bada",
  Tizen: "Tizen",
  Linux: "Linux",
  ChromeOS: "Chrome OS",
  PlayStation4: "PlayStation 4",
  Roku: "Roku"
}, P = {
  EdgeHTML: "EdgeHTML",
  Blink: "Blink",
  Trident: "Trident",
  Presto: "Presto",
  Gecko: "Gecko",
  WebKit: "WebKit"
};
class i {
  /**
   * Get first matched item for a string
   * @param {RegExp} regexp
   * @param {String} ua
   * @return {Array|{index: number, input: string}|*|boolean|string}
   */
  static getFirstMatch(e, r) {
    const n = r.match(e);
    return n && n.length > 0 && n[1] || "";
  }
  /**
   * Get second matched item for a string
   * @param regexp
   * @param {String} ua
   * @return {Array|{index: number, input: string}|*|boolean|string}
   */
  static getSecondMatch(e, r) {
    const n = r.match(e);
    return n && n.length > 1 && n[2] || "";
  }
  /**
   * Match a regexp and return a constant or undefined
   * @param {RegExp} regexp
   * @param {String} ua
   * @param {*} _const Any const that will be returned if regexp matches the string
   * @return {*}
   */
  static matchAndReturnConst(e, r, n) {
    if (e.test(r))
      return n;
  }
  static getWindowsVersionName(e) {
    switch (e) {
      case "NT":
        return "NT";
      case "XP":
        return "XP";
      case "NT 5.0":
        return "2000";
      case "NT 5.1":
        return "XP";
      case "NT 5.2":
        return "2003";
      case "NT 6.0":
        return "Vista";
      case "NT 6.1":
        return "7";
      case "NT 6.2":
        return "8";
      case "NT 6.3":
        return "8.1";
      case "NT 10.0":
        return "10";
      default:
        return;
    }
  }
  /**
   * Get macOS version name
   *    10.5 - Leopard
   *    10.6 - Snow Leopard
   *    10.7 - Lion
   *    10.8 - Mountain Lion
   *    10.9 - Mavericks
   *    10.10 - Yosemite
   *    10.11 - El Capitan
   *    10.12 - Sierra
   *    10.13 - High Sierra
   *    10.14 - Mojave
   *    10.15 - Catalina
   *
   * @example
   *   getMacOSVersionName("10.14") // 'Mojave'
   *
   * @param  {string} version
   * @return {string} versionName
   */
  static getMacOSVersionName(e) {
    const r = e.split(".").splice(0, 2).map((n) => parseInt(n, 10) || 0);
    if (r.push(0), r[0] === 10)
      switch (r[1]) {
        case 5:
          return "Leopard";
        case 6:
          return "Snow Leopard";
        case 7:
          return "Lion";
        case 8:
          return "Mountain Lion";
        case 9:
          return "Mavericks";
        case 10:
          return "Yosemite";
        case 11:
          return "El Capitan";
        case 12:
          return "Sierra";
        case 13:
          return "High Sierra";
        case 14:
          return "Mojave";
        case 15:
          return "Catalina";
        default:
          return;
      }
  }
  /**
   * Get Android version name
   *    1.5 - Cupcake
   *    1.6 - Donut
   *    2.0 - Eclair
   *    2.1 - Eclair
   *    2.2 - Froyo
   *    2.x - Gingerbread
   *    3.x - Honeycomb
   *    4.0 - Ice Cream Sandwich
   *    4.1 - Jelly Bean
   *    4.4 - KitKat
   *    5.x - Lollipop
   *    6.x - Marshmallow
   *    7.x - Nougat
   *    8.x - Oreo
   *    9.x - Pie
   *
   * @example
   *   getAndroidVersionName("7.0") // 'Nougat'
   *
   * @param  {string} version
   * @return {string} versionName
   */
  static getAndroidVersionName(e) {
    const r = e.split(".").splice(0, 2).map((n) => parseInt(n, 10) || 0);
    if (r.push(0), !(r[0] === 1 && r[1] < 5)) {
      if (r[0] === 1 && r[1] < 6)
        return "Cupcake";
      if (r[0] === 1 && r[1] >= 6)
        return "Donut";
      if (r[0] === 2 && r[1] < 2)
        return "Eclair";
      if (r[0] === 2 && r[1] === 2)
        return "Froyo";
      if (r[0] === 2 && r[1] > 2)
        return "Gingerbread";
      if (r[0] === 3)
        return "Honeycomb";
      if (r[0] === 4 && r[1] < 1)
        return "Ice Cream Sandwich";
      if (r[0] === 4 && r[1] < 4)
        return "Jelly Bean";
      if (r[0] === 4 && r[1] >= 4)
        return "KitKat";
      if (r[0] === 5)
        return "Lollipop";
      if (r[0] === 6)
        return "Marshmallow";
      if (r[0] === 7)
        return "Nougat";
      if (r[0] === 8)
        return "Oreo";
      if (r[0] === 9)
        return "Pie";
    }
  }
  /**
   * Get version precisions count
   *
   * @example
   *   getVersionPrecision("1.10.3") // 3
   *
   * @param  {string} version
   * @return {number}
   */
  static getVersionPrecision(e) {
    return e.split(".").length;
  }
  /**
   * Calculate browser version weight
   *
   * @example
   *   compareVersions('1.10.2.1',  '1.8.2.1.90')    // 1
   *   compareVersions('1.010.2.1', '1.09.2.1.90');  // 1
   *   compareVersions('1.10.2.1',  '1.10.2.1');     // 0
   *   compareVersions('1.10.2.1',  '1.0800.2');     // -1
   *   compareVersions('1.10.2.1',  '1.10',  true);  // 0
   *
   * @param {String} versionA versions versions to compare
   * @param {String} versionB versions versions to compare
   * @param {boolean} [isLoose] enable loose comparison
   * @return {Number} comparison result: -1 when versionA is lower,
   * 1 when versionA is bigger, 0 when both equal
   */
  /* eslint consistent-return: 1 */
  static compareVersions(e, r, n = !1) {
    const o = i.getVersionPrecision(e), s = i.getVersionPrecision(r);
    let d = Math.max(o, s), h = 0;
    const w = i.map([e, r], (l) => {
      const a = d - i.getVersionPrecision(l), g = l + new Array(a + 1).join(".0");
      return i.map(g.split("."), (y) => new Array(20 - y.length).join("0") + y).reverse();
    });
    for (n && (h = d - Math.min(o, s)), d -= 1; d >= h; ) {
      if (w[0][d] > w[1][d])
        return 1;
      if (w[0][d] === w[1][d]) {
        if (d === h)
          return 0;
        d -= 1;
      } else if (w[0][d] < w[1][d])
        return -1;
    }
  }
  /**
   * Array::map polyfill
   *
   * @param  {Array} arr
   * @param  {Function} iterator
   * @return {Array}
   */
  static map(e, r) {
    const n = [];
    let o;
    if (Array.prototype.map)
      return Array.prototype.map.call(e, r);
    for (o = 0; o < e.length; o += 1)
      n.push(r(e[o]));
    return n;
  }
  /**
   * Array::find polyfill
   *
   * @param  {Array} arr
   * @param  {Function} predicate
   * @return {Array}
   */
  static find(e, r) {
    let n, o;
    if (Array.prototype.find)
      return Array.prototype.find.call(e, r);
    for (n = 0, o = e.length; n < o; n += 1) {
      const s = e[n];
      if (r(s, n))
        return s;
    }
  }
  /**
   * Object::assign polyfill
   *
   * @param  {Object} obj
   * @param  {Object} ...objs
   * @return {Object}
   */
  static assign(e, ...r) {
    const n = e;
    let o, s;
    if (Object.assign)
      return Object.assign(e, ...r);
    for (o = 0, s = r.length; o < s; o += 1) {
      const d = r[o];
      typeof d == "object" && d !== null && Object.keys(d).forEach((w) => {
        n[w] = d[w];
      });
    }
    return e;
  }
  /**
   * Get short version/alias for a browser name
   *
   * @example
   *   getBrowserAlias('Microsoft Edge') // edge
   *
   * @param  {string} browserName
   * @return {string}
   */
  static getBrowserAlias(e) {
    return Ve[e];
  }
  /**
   * Get short version/alias for a browser name
   *
   * @example
   *   getBrowserAlias('edge') // Microsoft Edge
   *
   * @param  {string} browserAlias
   * @return {string}
   */
  static getBrowserTypeByAlias(e) {
    return be[e] || "";
  }
}
const k = /version\/(\d+(\.?_?\d+)+)/i, De = [
  /* Googlebot */
  {
    test: [/googlebot/i],
    describe(t) {
      const e = {
        name: "Googlebot"
      }, r = i.getFirstMatch(/googlebot\/(\d+(\.\d+))/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  /* Opera < 13.0 */
  {
    test: [/opera/i],
    describe(t) {
      const e = {
        name: "Opera"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:opera)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  /* Opera > 13.0 */
  {
    test: [/opr\/|opios/i],
    describe(t) {
      const e = {
        name: "Opera"
      }, r = i.getFirstMatch(/(?:opr|opios)[\s/](\S+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/SamsungBrowser/i],
    describe(t) {
      const e = {
        name: "Samsung Internet for Android"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:SamsungBrowser)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/Whale/i],
    describe(t) {
      const e = {
        name: "NAVER Whale Browser"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:whale)[\s/](\d+(?:\.\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/MZBrowser/i],
    describe(t) {
      const e = {
        name: "MZ Browser"
      }, r = i.getFirstMatch(/(?:MZBrowser)[\s/](\d+(?:\.\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/focus/i],
    describe(t) {
      const e = {
        name: "Focus"
      }, r = i.getFirstMatch(/(?:focus)[\s/](\d+(?:\.\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/swing/i],
    describe(t) {
      const e = {
        name: "Swing"
      }, r = i.getFirstMatch(/(?:swing)[\s/](\d+(?:\.\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/coast/i],
    describe(t) {
      const e = {
        name: "Opera Coast"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:coast)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/opt\/\d+(?:.?_?\d+)+/i],
    describe(t) {
      const e = {
        name: "Opera Touch"
      }, r = i.getFirstMatch(/(?:opt)[\s/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/yabrowser/i],
    describe(t) {
      const e = {
        name: "Yandex Browser"
      }, r = i.getFirstMatch(/(?:yabrowser)[\s/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/ucbrowser/i],
    describe(t) {
      const e = {
        name: "UC Browser"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:ucbrowser)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/Maxthon|mxios/i],
    describe(t) {
      const e = {
        name: "Maxthon"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:Maxthon|mxios)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/epiphany/i],
    describe(t) {
      const e = {
        name: "Epiphany"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:epiphany)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/puffin/i],
    describe(t) {
      const e = {
        name: "Puffin"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:puffin)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/sleipnir/i],
    describe(t) {
      const e = {
        name: "Sleipnir"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:sleipnir)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/k-meleon/i],
    describe(t) {
      const e = {
        name: "K-Meleon"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/(?:k-meleon)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/micromessenger/i],
    describe(t) {
      const e = {
        name: "WeChat"
      }, r = i.getFirstMatch(/(?:micromessenger)[\s/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/qqbrowser/i],
    describe(t) {
      const e = {
        name: /qqbrowserlite/i.test(t) ? "QQ Browser Lite" : "QQ Browser"
      }, r = i.getFirstMatch(/(?:qqbrowserlite|qqbrowser)[/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/msie|trident/i],
    describe(t) {
      const e = {
        name: "Internet Explorer"
      }, r = i.getFirstMatch(/(?:msie |rv:)(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/\sedg\//i],
    describe(t) {
      const e = {
        name: "Microsoft Edge"
      }, r = i.getFirstMatch(/\sedg\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/edg([ea]|ios)/i],
    describe(t) {
      const e = {
        name: "Microsoft Edge"
      }, r = i.getSecondMatch(/edg([ea]|ios)\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/vivaldi/i],
    describe(t) {
      const e = {
        name: "Vivaldi"
      }, r = i.getFirstMatch(/vivaldi\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/seamonkey/i],
    describe(t) {
      const e = {
        name: "SeaMonkey"
      }, r = i.getFirstMatch(/seamonkey\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/sailfish/i],
    describe(t) {
      const e = {
        name: "Sailfish"
      }, r = i.getFirstMatch(/sailfish\s?browser\/(\d+(\.\d+)?)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/silk/i],
    describe(t) {
      const e = {
        name: "Amazon Silk"
      }, r = i.getFirstMatch(/silk\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/phantom/i],
    describe(t) {
      const e = {
        name: "PhantomJS"
      }, r = i.getFirstMatch(/phantomjs\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/slimerjs/i],
    describe(t) {
      const e = {
        name: "SlimerJS"
      }, r = i.getFirstMatch(/slimerjs\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/blackberry|\bbb\d+/i, /rim\stablet/i],
    describe(t) {
      const e = {
        name: "BlackBerry"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/blackberry[\d]+\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/(web|hpw)[o0]s/i],
    describe(t) {
      const e = {
        name: "WebOS Browser"
      }, r = i.getFirstMatch(k, t) || i.getFirstMatch(/w(?:eb)?[o0]sbrowser\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/bada/i],
    describe(t) {
      const e = {
        name: "Bada"
      }, r = i.getFirstMatch(/dolfin\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/tizen/i],
    describe(t) {
      const e = {
        name: "Tizen"
      }, r = i.getFirstMatch(/(?:tizen\s?)?browser\/(\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/qupzilla/i],
    describe(t) {
      const e = {
        name: "QupZilla"
      }, r = i.getFirstMatch(/(?:qupzilla)[\s/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/firefox|iceweasel|fxios/i],
    describe(t) {
      const e = {
        name: "Firefox"
      }, r = i.getFirstMatch(/(?:firefox|iceweasel|fxios)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/electron/i],
    describe(t) {
      const e = {
        name: "Electron"
      }, r = i.getFirstMatch(/(?:electron)\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/MiuiBrowser/i],
    describe(t) {
      const e = {
        name: "Miui"
      }, r = i.getFirstMatch(/(?:MiuiBrowser)[\s/](\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/chromium/i],
    describe(t) {
      const e = {
        name: "Chromium"
      }, r = i.getFirstMatch(/(?:chromium)[\s/](\d+(\.?_?\d+)+)/i, t) || i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/chrome|crios|crmo/i],
    describe(t) {
      const e = {
        name: "Chrome"
      }, r = i.getFirstMatch(/(?:chrome|crios|crmo)\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  {
    test: [/GSA/i],
    describe(t) {
      const e = {
        name: "Google Search"
      }, r = i.getFirstMatch(/(?:GSA)\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  /* Android Browser */
  {
    test(t) {
      const e = !t.test(/like android/i), r = t.test(/android/i);
      return e && r;
    },
    describe(t) {
      const e = {
        name: "Android Browser"
      }, r = i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  /* PlayStation 4 */
  {
    test: [/playstation 4/i],
    describe(t) {
      const e = {
        name: "PlayStation 4"
      }, r = i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  /* Safari */
  {
    test: [/safari|applewebkit/i],
    describe(t) {
      const e = {
        name: "Safari"
      }, r = i.getFirstMatch(k, t);
      return r && (e.version = r), e;
    }
  },
  /* Something else */
  {
    test: [/.*/i],
    describe(t) {
      const e = /^(.*)\/(.*) /, r = /^(.*)\/(.*)[ \t]\((.*)/, o = t.search("\\(") !== -1 ? r : e;
      return {
        name: i.getFirstMatch(o, t),
        version: i.getSecondMatch(o, t)
      };
    }
  }
], qe = [
  /* Roku */
  {
    test: [/Roku\/DVP/],
    describe(t) {
      const e = i.getFirstMatch(/Roku\/DVP-(\d+\.\d+)/i, t);
      return {
        name: F.Roku,
        version: e
      };
    }
  },
  /* Windows Phone */
  {
    test: [/windows phone/i],
    describe(t) {
      const e = i.getFirstMatch(/windows phone (?:os)?\s?(\d+(\.\d+)*)/i, t);
      return {
        name: F.WindowsPhone,
        version: e
      };
    }
  },
  /* Windows */
  {
    test: [/windows /i],
    describe(t) {
      const e = i.getFirstMatch(/Windows ((NT|XP)( \d\d?.\d)?)/i, t), r = i.getWindowsVersionName(e);
      return {
        name: F.Windows,
        version: e,
        versionName: r
      };
    }
  },
  /* Firefox on iPad */
  {
    test: [/Macintosh(.*?) FxiOS(.*?)\//],
    describe(t) {
      const e = {
        name: F.iOS
      }, r = i.getSecondMatch(/(Version\/)(\d[\d.]+)/, t);
      return r && (e.version = r), e;
    }
  },
  /* macOS */
  {
    test: [/macintosh/i],
    describe(t) {
      const e = i.getFirstMatch(/mac os x (\d+(\.?_?\d+)+)/i, t).replace(/[_\s]/g, "."), r = i.getMacOSVersionName(e), n = {
        name: F.MacOS,
        version: e
      };
      return r && (n.versionName = r), n;
    }
  },
  /* iOS */
  {
    test: [/(ipod|iphone|ipad)/i],
    describe(t) {
      const e = i.getFirstMatch(/os (\d+([_\s]\d+)*) like mac os x/i, t).replace(/[_\s]/g, ".");
      return {
        name: F.iOS,
        version: e
      };
    }
  },
  /* Android */
  {
    test(t) {
      const e = !t.test(/like android/i), r = t.test(/android/i);
      return e && r;
    },
    describe(t) {
      const e = i.getFirstMatch(/android[\s/-](\d+(\.\d+)*)/i, t), r = i.getAndroidVersionName(e), n = {
        name: F.Android,
        version: e
      };
      return r && (n.versionName = r), n;
    }
  },
  /* WebOS */
  {
    test: [/(web|hpw)[o0]s/i],
    describe(t) {
      const e = i.getFirstMatch(/(?:web|hpw)[o0]s\/(\d+(\.\d+)*)/i, t), r = {
        name: F.WebOS
      };
      return e && e.length && (r.version = e), r;
    }
  },
  /* BlackBerry */
  {
    test: [/blackberry|\bbb\d+/i, /rim\stablet/i],
    describe(t) {
      const e = i.getFirstMatch(/rim\stablet\sos\s(\d+(\.\d+)*)/i, t) || i.getFirstMatch(/blackberry\d+\/(\d+([_\s]\d+)*)/i, t) || i.getFirstMatch(/\bbb(\d+)/i, t);
      return {
        name: F.BlackBerry,
        version: e
      };
    }
  },
  /* Bada */
  {
    test: [/bada/i],
    describe(t) {
      const e = i.getFirstMatch(/bada\/(\d+(\.\d+)*)/i, t);
      return {
        name: F.Bada,
        version: e
      };
    }
  },
  /* Tizen */
  {
    test: [/tizen/i],
    describe(t) {
      const e = i.getFirstMatch(/tizen[/\s](\d+(\.\d+)*)/i, t);
      return {
        name: F.Tizen,
        version: e
      };
    }
  },
  /* Linux */
  {
    test: [/linux/i],
    describe() {
      return {
        name: F.Linux
      };
    }
  },
  /* Chrome OS */
  {
    test: [/CrOS/],
    describe() {
      return {
        name: F.ChromeOS
      };
    }
  },
  /* Playstation 4 */
  {
    test: [/PlayStation 4/],
    describe(t) {
      const e = i.getFirstMatch(/PlayStation 4[/\s](\d+(\.\d+)*)/i, t);
      return {
        name: F.PlayStation4,
        version: e
      };
    }
  }
], Ge = [
  /* Googlebot */
  {
    test: [/googlebot/i],
    describe() {
      return {
        type: "bot",
        vendor: "Google"
      };
    }
  },
  /* Huawei */
  {
    test: [/huawei/i],
    describe(t) {
      const e = i.getFirstMatch(/(can-l01)/i, t) && "Nova", r = {
        type: M.mobile,
        vendor: "Huawei"
      };
      return e && (r.model = e), r;
    }
  },
  /* Nexus Tablet */
  {
    test: [/nexus\s*(?:7|8|9|10).*/i],
    describe() {
      return {
        type: M.tablet,
        vendor: "Nexus"
      };
    }
  },
  /* iPad */
  {
    test: [/ipad/i],
    describe() {
      return {
        type: M.tablet,
        vendor: "Apple",
        model: "iPad"
      };
    }
  },
  /* Firefox on iPad */
  {
    test: [/Macintosh(.*?) FxiOS(.*?)\//],
    describe() {
      return {
        type: M.tablet,
        vendor: "Apple",
        model: "iPad"
      };
    }
  },
  /* Amazon Kindle Fire */
  {
    test: [/kftt build/i],
    describe() {
      return {
        type: M.tablet,
        vendor: "Amazon",
        model: "Kindle Fire HD 7"
      };
    }
  },
  /* Another Amazon Tablet with Silk */
  {
    test: [/silk/i],
    describe() {
      return {
        type: M.tablet,
        vendor: "Amazon"
      };
    }
  },
  /* Tablet */
  {
    test: [/tablet(?! pc)/i],
    describe() {
      return {
        type: M.tablet
      };
    }
  },
  /* iPod/iPhone */
  {
    test(t) {
      const e = t.test(/ipod|iphone/i), r = t.test(/like (ipod|iphone)/i);
      return e && !r;
    },
    describe(t) {
      const e = i.getFirstMatch(/(ipod|iphone)/i, t);
      return {
        type: M.mobile,
        vendor: "Apple",
        model: e
      };
    }
  },
  /* Nexus Mobile */
  {
    test: [/nexus\s*[0-6].*/i, /galaxy nexus/i],
    describe() {
      return {
        type: M.mobile,
        vendor: "Nexus"
      };
    }
  },
  /* Mobile */
  {
    test: [/[^-]mobi/i],
    describe() {
      return {
        type: M.mobile
      };
    }
  },
  /* BlackBerry */
  {
    test(t) {
      return t.getBrowserName(!0) === "blackberry";
    },
    describe() {
      return {
        type: M.mobile,
        vendor: "BlackBerry"
      };
    }
  },
  /* Bada */
  {
    test(t) {
      return t.getBrowserName(!0) === "bada";
    },
    describe() {
      return {
        type: M.mobile
      };
    }
  },
  /* Windows Phone */
  {
    test(t) {
      return t.getBrowserName() === "windows phone";
    },
    describe() {
      return {
        type: M.mobile,
        vendor: "Microsoft"
      };
    }
  },
  /* Android Tablet */
  {
    test(t) {
      const e = Number(String(t.getOSVersion()).split(".")[0]);
      return t.getOSName(!0) === "android" && e >= 3;
    },
    describe() {
      return {
        type: M.tablet
      };
    }
  },
  /* Android Mobile */
  {
    test(t) {
      return t.getOSName(!0) === "android";
    },
    describe() {
      return {
        type: M.mobile
      };
    }
  },
  /* desktop */
  {
    test(t) {
      return t.getOSName(!0) === "macos";
    },
    describe() {
      return {
        type: M.desktop,
        vendor: "Apple"
      };
    }
  },
  /* Windows */
  {
    test(t) {
      return t.getOSName(!0) === "windows";
    },
    describe() {
      return {
        type: M.desktop
      };
    }
  },
  /* Linux */
  {
    test(t) {
      return t.getOSName(!0) === "linux";
    },
    describe() {
      return {
        type: M.desktop
      };
    }
  },
  /* PlayStation 4 */
  {
    test(t) {
      return t.getOSName(!0) === "playstation 4";
    },
    describe() {
      return {
        type: M.tv
      };
    }
  },
  /* Roku */
  {
    test(t) {
      return t.getOSName(!0) === "roku";
    },
    describe() {
      return {
        type: M.tv
      };
    }
  }
], Qe = [
  /* EdgeHTML */
  {
    test(t) {
      return t.getBrowserName(!0) === "microsoft edge";
    },
    describe(t) {
      if (/\sedg\//i.test(t))
        return {
          name: P.Blink
        };
      const r = i.getFirstMatch(/edge\/(\d+(\.?_?\d+)+)/i, t);
      return {
        name: P.EdgeHTML,
        version: r
      };
    }
  },
  /* Trident */
  {
    test: [/trident/i],
    describe(t) {
      const e = {
        name: P.Trident
      }, r = i.getFirstMatch(/trident\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  /* Presto */
  {
    test(t) {
      return t.test(/presto/i);
    },
    describe(t) {
      const e = {
        name: P.Presto
      }, r = i.getFirstMatch(/presto\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  /* Gecko */
  {
    test(t) {
      const e = t.test(/gecko/i), r = t.test(/like gecko/i);
      return e && !r;
    },
    describe(t) {
      const e = {
        name: P.Gecko
      }, r = i.getFirstMatch(/gecko\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  },
  /* Blink */
  {
    test: [/(apple)?webkit\/537\.36/i],
    describe() {
      return {
        name: P.Blink
      };
    }
  },
  /* WebKit */
  {
    test: [/(apple)?webkit/i],
    describe(t) {
      const e = {
        name: P.WebKit
      }, r = i.getFirstMatch(/webkit\/(\d+(\.?_?\d+)+)/i, t);
      return r && (e.version = r), e;
    }
  }
];
class se {
  /**
   * Create instance of Parser
   *
   * @param {String} UA User-Agent string
   * @param {Boolean} [skipParsing=false] parser can skip parsing in purpose of performance
   * improvements if you need to make a more particular parsing
   * like {@link Parser#parseBrowser} or {@link Parser#parsePlatform}
   *
   * @throw {Error} in case of empty UA String
   *
   * @constructor
   */
  constructor(e, r = !1) {
    if (e == null || e === "")
      throw new Error("UserAgent parameter can't be empty");
    this._ua = e, this.parsedResult = {}, r !== !0 && this.parse();
  }
  /**
   * Get UserAgent string of current Parser instance
   * @return {String} User-Agent String of the current <Parser> object
   *
   * @public
   */
  getUA() {
    return this._ua;
  }
  /**
   * Test a UA string for a regexp
   * @param {RegExp} regex
   * @return {Boolean}
   */
  test(e) {
    return e.test(this._ua);
  }
  /**
   * Get parsed browser object
   * @return {Object}
   */
  parseBrowser() {
    this.parsedResult.browser = {};
    const e = i.find(De, (r) => {
      if (typeof r.test == "function")
        return r.test(this);
      if (r.test instanceof Array)
        return r.test.some((n) => this.test(n));
      throw new Error("Browser's test function is not valid");
    });
    return e && (this.parsedResult.browser = e.describe(this.getUA())), this.parsedResult.browser;
  }
  /**
   * Get parsed browser object
   * @return {Object}
   *
   * @public
   */
  getBrowser() {
    return this.parsedResult.browser ? this.parsedResult.browser : this.parseBrowser();
  }
  /**
   * Get browser's name
   * @return {String} Browser's name or an empty string
   *
   * @public
   */
  getBrowserName(e) {
    return e ? String(this.getBrowser().name).toLowerCase() || "" : this.getBrowser().name || "";
  }
  /**
   * Get browser's version
   * @return {String} version of browser
   *
   * @public
   */
  getBrowserVersion() {
    return this.getBrowser().version;
  }
  /**
   * Get OS
   * @return {Object}
   *
   * @example
   * this.getOS();
   * {
   *   name: 'macOS',
   *   version: '10.11.12'
   * }
   */
  getOS() {
    return this.parsedResult.os ? this.parsedResult.os : this.parseOS();
  }
  /**
   * Parse OS and save it to this.parsedResult.os
   * @return {*|{}}
   */
  parseOS() {
    this.parsedResult.os = {};
    const e = i.find(qe, (r) => {
      if (typeof r.test == "function")
        return r.test(this);
      if (r.test instanceof Array)
        return r.test.some((n) => this.test(n));
      throw new Error("Browser's test function is not valid");
    });
    return e && (this.parsedResult.os = e.describe(this.getUA())), this.parsedResult.os;
  }
  /**
   * Get OS name
   * @param {Boolean} [toLowerCase] return lower-cased value
   * @return {String} name of the OS — macOS, Windows, Linux, etc.
   */
  getOSName(e) {
    const { name: r } = this.getOS();
    return e ? String(r).toLowerCase() || "" : r || "";
  }
  /**
   * Get OS version
   * @return {String} full version with dots ('10.11.12', '5.6', etc)
   */
  getOSVersion() {
    return this.getOS().version;
  }
  /**
   * Get parsed platform
   * @return {{}}
   */
  getPlatform() {
    return this.parsedResult.platform ? this.parsedResult.platform : this.parsePlatform();
  }
  /**
   * Get platform name
   * @param {Boolean} [toLowerCase=false]
   * @return {*}
   */
  getPlatformType(e = !1) {
    const { type: r } = this.getPlatform();
    return e ? String(r).toLowerCase() || "" : r || "";
  }
  /**
   * Get parsed platform
   * @return {{}}
   */
  parsePlatform() {
    this.parsedResult.platform = {};
    const e = i.find(Ge, (r) => {
      if (typeof r.test == "function")
        return r.test(this);
      if (r.test instanceof Array)
        return r.test.some((n) => this.test(n));
      throw new Error("Browser's test function is not valid");
    });
    return e && (this.parsedResult.platform = e.describe(this.getUA())), this.parsedResult.platform;
  }
  /**
   * Get parsed engine
   * @return {{}}
   */
  getEngine() {
    return this.parsedResult.engine ? this.parsedResult.engine : this.parseEngine();
  }
  /**
   * Get engines's name
   * @return {String} Engines's name or an empty string
   *
   * @public
   */
  getEngineName(e) {
    return e ? String(this.getEngine().name).toLowerCase() || "" : this.getEngine().name || "";
  }
  /**
   * Get parsed platform
   * @return {{}}
   */
  parseEngine() {
    this.parsedResult.engine = {};
    const e = i.find(Qe, (r) => {
      if (typeof r.test == "function")
        return r.test(this);
      if (r.test instanceof Array)
        return r.test.some((n) => this.test(n));
      throw new Error("Browser's test function is not valid");
    });
    return e && (this.parsedResult.engine = e.describe(this.getUA())), this.parsedResult.engine;
  }
  /**
   * Parse full information about the browser
   * @returns {Parser}
   */
  parse() {
    return this.parseBrowser(), this.parseOS(), this.parsePlatform(), this.parseEngine(), this;
  }
  /**
   * Get parsed result
   * @return {ParsedResult}
   */
  getResult() {
    return i.assign({}, this.parsedResult);
  }
  /**
   * Check if parsed browser matches certain conditions
   *
   * @param {Object} checkTree It's one or two layered object,
   * which can include a platform or an OS on the first layer
   * and should have browsers specs on the bottom-laying layer
   *
   * @returns {Boolean|undefined} Whether the browser satisfies the set conditions or not.
   * Returns `undefined` when the browser is no described in the checkTree object.
   *
   * @example
   * const browser = Bowser.getParser(window.navigator.userAgent);
   * if (browser.satisfies({chrome: '>118.01.1322' }))
   * // or with os
   * if (browser.satisfies({windows: { chrome: '>118.01.1322' } }))
   * // or with platforms
   * if (browser.satisfies({desktop: { chrome: '>118.01.1322' } }))
   */
  satisfies(e) {
    const r = {};
    let n = 0;
    const o = {};
    let s = 0;
    if (Object.keys(e).forEach((h) => {
      const w = e[h];
      typeof w == "string" ? (o[h] = w, s += 1) : typeof w == "object" && (r[h] = w, n += 1);
    }), n > 0) {
      const h = Object.keys(r), w = i.find(h, (a) => this.isOS(a));
      if (w) {
        const a = this.satisfies(r[w]);
        if (a !== void 0)
          return a;
      }
      const l = i.find(
        h,
        (a) => this.isPlatform(a)
      );
      if (l) {
        const a = this.satisfies(r[l]);
        if (a !== void 0)
          return a;
      }
    }
    if (s > 0) {
      const h = Object.keys(o), w = i.find(h, (l) => this.isBrowser(l, !0));
      if (w !== void 0)
        return this.compareVersion(o[w]);
    }
  }
  /**
   * Check if the browser name equals the passed string
   * @param browserName The string to compare with the browser name
   * @param [includingAlias=false] The flag showing whether alias will be included into comparison
   * @returns {boolean}
   */
  isBrowser(e, r = !1) {
    const n = this.getBrowserName().toLowerCase();
    let o = e.toLowerCase();
    const s = i.getBrowserTypeByAlias(o);
    return r && s && (o = s.toLowerCase()), o === n;
  }
  compareVersion(e) {
    let r = [0], n = e, o = !1;
    const s = this.getBrowserVersion();
    if (typeof s == "string")
      return e[0] === ">" || e[0] === "<" ? (n = e.substr(1), e[1] === "=" ? (o = !0, n = e.substr(2)) : r = [], e[0] === ">" ? r.push(1) : r.push(-1)) : e[0] === "=" ? n = e.substr(1) : e[0] === "~" && (o = !0, n = e.substr(1)), r.indexOf(
        i.compareVersions(s, n, o)
      ) > -1;
  }
  isOS(e) {
    return this.getOSName(!0) === String(e).toLowerCase();
  }
  isPlatform(e) {
    return this.getPlatformType(!0) === String(e).toLowerCase();
  }
  isEngine(e) {
    return this.getEngineName(!0) === String(e).toLowerCase();
  }
  /**
   * Is anything? Check if the browser is called "anything",
   * the OS called "anything" or the platform called "anything"
   * @param {String} anything
   * @param [includingAlias=false] The flag showing whether alias will be included into comparison
   * @returns {Boolean}
   */
  is(e, r = !1) {
    return this.isBrowser(e, r) || this.isOS(e) || this.isPlatform(e);
  }
  /**
   * Check if any of the given values satisfies this.is(anything)
   * @param {String[]} anythings
   * @returns {Boolean}
   */
  some(e = []) {
    return e.some((r) => this.is(r));
  }
}
/*!
 * Bowser - a browser detector
 * https://github.com/lancedikson/bowser
 * MIT License | (c) Dustin Diaz 2012-2015
 * MIT License | (c) Denis Demchenko 2015-2019
 */
class Xe {
  /**
   * Creates a {@link Parser} instance
   *
   * @param {String} UA UserAgent string
   * @param {Boolean} [skipParsing=false] Will make the Parser postpone parsing until you ask it
   * explicitly. Same as `skipParsing` for {@link Parser}.
   * @returns {Parser}
   * @throws {Error} when UA is not a String
   *
   * @example
   * const parser = Bowser.getParser(window.navigator.userAgent);
   * const result = parser.getResult();
   */
  static getParser(e, r = !1) {
    if (typeof e != "string")
      throw new Error("UserAgent should be a string");
    return new se(e, r);
  }
  /**
   * Creates a {@link Parser} instance and runs {@link Parser.getResult} immediately
   *
   * @param UA
   * @return {ParsedResult}
   *
   * @example
   * const result = Bowser.parse(window.navigator.userAgent);
   */
  static parse(e) {
    return new se(e).getResult();
  }
  static get BROWSER_MAP() {
    return be;
  }
  static get ENGINE_MAP() {
    return P;
  }
  static get OS_MAP() {
    return F;
  }
  static get PLATFORMS_MAP() {
    return M;
  }
}
const ae = typeof window < "u" ? window : null;
function He() {
  if (!ae)
    return null;
  switch (Xe.getParser(ae.navigator.userAgent).getBrowserName()?.toLowerCase()) {
    case "firefox":
      return "firefox";
    case "microsoft edge":
      return "edge";
    case "android browser":
    case "chrome":
    case "chromium":
    case "electron":
    case "opera":
    case "vivaldi":
      return "chrome";
    default:
      return null;
  }
}
var Ue = typeof global == "object" && global && global.Object === Object && global;
const Ze = Ue;
var Ke = typeof self == "object" && self && self.Object === Object && self, Ye = Ze || Ke || Function("return this")();
const $e = Ye;
var Je = $e.Symbol;
const Y = Je;
var me = Object.prototype, et = me.hasOwnProperty, tt = me.toString, G = Y ? Y.toStringTag : void 0;
function rt(t) {
  var e = et.call(t, G), r = t[G];
  try {
    t[G] = void 0;
    var n = !0;
  } catch {
  }
  var o = tt.call(t);
  return n && (e ? t[G] = r : delete t[G]), o;
}
var nt = Object.prototype, it = nt.toString;
function ot(t) {
  return it.call(t);
}
var st = "[object Null]", at = "[object Undefined]", ce = Y ? Y.toStringTag : void 0;
function ct(t) {
  return t == null ? t === void 0 ? at : st : ce && ce in Object(t) ? rt(t) : ot(t);
}
function lt(t) {
  return t != null && typeof t == "object";
}
var dt = Array.isArray;
const ut = dt;
var ft = "[object String]";
function te(t) {
  return typeof t == "string" || !ut(t) && lt(t) && ct(t) == ft;
}
const wt = ({
  availableConnectors: t,
  installedWallets: e,
  discoveryWallets: r,
  storeVersion: n,
  customOrder: o
}) => {
  if (window?.starknet_argentX?.isInAppBrowser)
    return [];
  const h = e.map(
    (a) => t.find((g) => g.id === a.id)
  );
  return (o ? t : [
    ...t.filter((a) => h.includes(a)),
    ...t.filter((a) => !h.includes(a))
  ]).map((a) => {
    const g = e.find((f) => f.id === a.id);
    if (g) {
      const f = g.id === "argentX" ? { light: ee, dark: ee } : te(g.icon) ? { light: g.icon, dark: g.icon } : g.icon;
      return {
        name: g.name,
        id: g.id,
        icon: f,
        connector: a
      };
    }
    const y = r.filter(
      (f) => !!f.downloads[n]
    ).find((f) => f.id === a.id);
    if (y) {
      const { downloads: f } = y, b = y.id === "argentX" ? ee : y.icon;
      return {
        name: y.name,
        id: y.id,
        icon: { light: b, dark: b },
        connector: a,
        download: f[n]
      };
    }
    return !a || !a.id || !a.name ? null : {
      name: a.name,
      id: a.id,
      icon: a.icon,
      connector: a,
      title: "title" in a && te(a.title) ? a.title : void 0,
      subtitle: "subtitle" in a && te(a.subtitle) ? a.subtitle : void 0
    };
  }).filter((a) => a !== null);
};
function R() {
}
function ve(t) {
  return t();
}
function le() {
  return /* @__PURE__ */ Object.create(null);
}
function j(t) {
  t.forEach(ve);
}
function ye(t) {
  return typeof t == "function";
}
function ke(t, e) {
  return t != t ? e == e : t !== e || t && typeof t == "object" || typeof t == "function";
}
let U;
function _e(t, e) {
  return t === e ? !0 : (U || (U = document.createElement("a")), U.href = e, t === U.href);
}
function gt(t) {
  return Object.keys(t).length === 0;
}
function _(t, e) {
  t.appendChild(e);
}
function W(t, e, r) {
  t.insertBefore(e, r || null);
}
function E(t) {
  t.parentNode && t.parentNode.removeChild(t);
}
function ht(t, e) {
  for (let r = 0; r < t.length; r += 1)
    t[r] && t[r].d(e);
}
function B(t) {
  return document.createElement(t);
}
function I(t) {
  return document.createTextNode(t);
}
function z() {
  return I(" ");
}
function xe() {
  return I("");
}
function O(t, e, r, n) {
  return t.addEventListener(e, r, n), () => t.removeEventListener(e, r, n);
}
function p(t, e, r) {
  r == null ? t.removeAttribute(e) : t.getAttribute(e) !== r && t.setAttribute(e, r);
}
function pt(t) {
  return Array.from(t.childNodes);
}
function $(t, e) {
  e = "" + e, t.data !== e && (t.data = /** @type {string} */
  e);
}
function Me(t, e, r, n) {
  r == null ? t.style.removeProperty(e) : t.style.setProperty(e, r, n ? "important" : "");
}
let X;
function Q(t) {
  X = t;
}
function bt() {
  if (!X)
    throw new Error("Function called outside component initialization");
  return X;
}
function mt(t) {
  bt().$$.on_mount.push(t);
}
const D = [], de = [];
let q = [];
const ue = [], vt = /* @__PURE__ */ Promise.resolve();
let ne = !1;
function yt() {
  ne || (ne = !0, vt.then(Se));
}
function ie(t) {
  q.push(t);
}
const re = /* @__PURE__ */ new Set();
let V = 0;
function Se() {
  if (V !== 0)
    return;
  const t = X;
  do {
    try {
      for (; V < D.length; ) {
        const e = D[V];
        V++, Q(e), kt(e.$$);
      }
    } catch (e) {
      throw D.length = 0, V = 0, e;
    }
    for (Q(null), D.length = 0, V = 0; de.length; )
      de.pop()();
    for (let e = 0; e < q.length; e += 1) {
      const r = q[e];
      re.has(r) || (re.add(r), r());
    }
    q.length = 0;
  } while (D.length);
  for (; ue.length; )
    ue.pop()();
  ne = !1, re.clear(), Q(t);
}
function kt(t) {
  if (t.fragment !== null) {
    t.update(), j(t.before_update);
    const e = t.dirty;
    t.dirty = [-1], t.fragment && t.fragment.p(t.ctx, e), t.after_update.forEach(ie);
  }
}
function _t(t) {
  const e = [], r = [];
  q.forEach((n) => t.indexOf(n) === -1 ? e.push(n) : r.push(n)), r.forEach((n) => n()), q = e;
}
const K = /* @__PURE__ */ new Set();
let T;
function Be() {
  T = {
    r: 0,
    c: [],
    p: T
    // parent group
  };
}
function Fe() {
  T.r || j(T.c), T = T.p;
}
function L(t, e) {
  t && t.i && (K.delete(t), t.i(e));
}
function H(t, e, r, n) {
  if (t && t.o) {
    if (K.has(t))
      return;
    K.add(t), T.c.push(() => {
      K.delete(t), n && (r && t.d(1), n());
    }), t.o(e);
  } else
    n && n();
}
function fe(t) {
  return t?.length !== void 0 ? t : Array.from(t);
}
function xt(t) {
  t && t.c();
}
function Ce(t, e, r) {
  const { fragment: n, after_update: o } = t.$$;
  n && n.m(e, r), ie(() => {
    const s = t.$$.on_mount.map(ve).filter(ye);
    t.$$.on_destroy ? t.$$.on_destroy.push(...s) : j(s), t.$$.on_mount = [];
  }), o.forEach(ie);
}
function Ae(t, e) {
  const r = t.$$;
  r.fragment !== null && (_t(r.after_update), j(r.on_destroy), r.fragment && r.fragment.d(e), r.on_destroy = r.fragment = null, r.ctx = []);
}
function Mt(t, e) {
  t.$$.dirty[0] === -1 && (D.push(t), yt(), t.$$.dirty.fill(0)), t.$$.dirty[e / 31 | 0] |= 1 << e % 31;
}
function Oe(t, e, r, n, o, s, d = null, h = [-1]) {
  const w = X;
  Q(t);
  const l = t.$$ = {
    fragment: null,
    ctx: [],
    // state
    props: s,
    update: R,
    not_equal: o,
    bound: le(),
    // lifecycle
    on_mount: [],
    on_destroy: [],
    on_disconnect: [],
    before_update: [],
    after_update: [],
    context: new Map(e.context || (w ? w.$$.context : [])),
    // everything else
    callbacks: le(),
    dirty: h,
    skip_bound: !1,
    root: e.target || w.$$.root
  };
  d && d(l.root);
  let a = !1;
  if (l.ctx = r ? r(t, e.props || {}, (g, y, ...f) => {
    const b = f.length ? f[0] : y;
    return l.ctx && o(l.ctx[g], l.ctx[g] = b) && (!l.skip_bound && l.bound[g] && l.bound[g](b), a && Mt(t, g)), y;
  }) : [], l.update(), a = !0, j(l.before_update), l.fragment = n ? n(l.ctx) : !1, e.target) {
    if (e.hydrate) {
      const g = pt(e.target);
      l.fragment && l.fragment.l(g), g.forEach(E);
    } else
      l.fragment && l.fragment.c();
    e.intro && L(t.$$.fragment), Ce(t, e.target, e.anchor), Se();
  }
  Q(w);
}
class Ne {
  constructor() {
    /**
     * ### PRIVATE API
     *
     * Do not use, may change at any time
     *
     * @type {any}
     */
    J(this, "$$");
    /**
     * ### PRIVATE API
     *
     * Do not use, may change at any time
     *
     * @type {any}
     */
    J(this, "$$set");
  }
  /** @returns {void} */
  $destroy() {
    Ae(this, 1), this.$destroy = R;
  }
  /**
   * @template {Extract<keyof Events, string>} K
   * @param {K} type
   * @param {((e: Events[K]) => void) | null | undefined} callback
   * @returns {() => void}
   */
  $on(e, r) {
    if (!ye(r))
      return R;
    const n = this.$$.callbacks[e] || (this.$$.callbacks[e] = []);
    return n.push(r), () => {
      const o = n.indexOf(r);
      o !== -1 && n.splice(o, 1);
    };
  }
  /**
   * @param {Partial<Props>} props
   * @returns {void}
   */
  $set(e) {
    this.$$set && !gt(e) && (this.$$.skip_bound = !0, this.$$set(e), this.$$.skip_bound = !1);
  }
}
const St = "4";
typeof window < "u" && (window.__svelte || (window.__svelte = { v: /* @__PURE__ */ new Set() })).v.add(St);
function Bt(t) {
  let e, r, n, o, s, d = (
    /*wallet*/
    (t[0].title ?? /*wallet*/
    t[0].name) + ""
  ), h, w, l, a = (
    /*wallet*/
    (t[0].subtitle ?? "") + ""
  ), g, y, f, b;
  function x(v, u) {
    return (
      /*loadingItem*/
      v[2] === /*wallet*/
      v[0]?.id ? Ot : (
        /*isSvg*/
        v[4] ? At : Ct
      )
    );
  }
  let m = x(t), c = m(t);
  return {
    c() {
      e = B("li"), r = B("span"), n = z(), o = B("div"), s = B("p"), h = I(d), w = z(), l = B("p"), g = I(a), y = z(), c.c(), p(r, "class", "w-8 h-8"), p(s, "class", "font-semibold text-base p"), p(l, "class", "l2 p"), Me(l, "text-align", "center"), p(o, "class", "flex flex-col justify-center items-center"), p(e, "class", `flex flex-row-reverse justify-between items-center 
            p-3 rounded-md cursor-pointer shadow-list-item 
            dark:shadow-none dark:bg-neutral-800 dark:text-white 
          hover:bg-neutral-100 dark:hover:bg-neutral-700 
          focus:outline-none focus:ring-2 
        focus:ring-neutral-200 dark:focus:ring-neutral-700 
          transition-colors`), p(e, "role", "button"), p(e, "tabindex", "0");
    },
    m(v, u) {
      W(v, e, u), _(e, r), _(e, n), _(e, o), _(o, s), _(s, h), _(o, w), _(o, l), _(l, g), _(e, y), c.m(e, null), f || (b = [
        O(
          e,
          "click",
          /*click_handler_1*/
          t[8]
        ),
        O(
          e,
          "keyup",
          /*keyup_handler_1*/
          t[9]
        )
      ], f = !0);
    },
    p(v, u) {
      u & /*wallet*/
      1 && d !== (d = /*wallet*/
      (v[0].title ?? /*wallet*/
      v[0].name) + "") && $(h, d), u & /*wallet*/
      1 && a !== (a = /*wallet*/
      (v[0].subtitle ?? "") + "") && $(g, a), m === (m = x(v)) && c ? c.p(v, u) : (c.d(1), c = m(v), c && (c.c(), c.m(e, null)));
    },
    d(v) {
      v && E(e), c.d(), f = !1, j(b);
    }
  };
}
function Ft(t) {
  let e, r, n, o, s, d, h = (
    /*wallet*/
    t[0].name + ""
  ), w, l, a, g, y, f, b, x, m;
  return {
    c() {
      e = B("a"), r = B("li"), n = B("span"), o = z(), s = B("p"), d = I("Install "), w = I(h), l = z(), a = B("img"), p(n, "class", "w-8 h-8"), p(s, "class", "font-semibold text-base p"), p(a, "alt", g = /*wallet*/
      t[0].name), _e(a.src, y = /*icon*/
      t[3]) || p(a, "src", y), p(a, "class", "w-8 h-8 rounded-full"), p(r, "class", `flex flex-row-reverse justify-between items-center 
              p-3 rounded-md cursor-pointer shadow-list-item 
              dark:shadow-none dark:bg-neutral-800 dark:text-white 
            hover:bg-neutral-100 dark:hover:bg-neutral-700`), p(e, "aria-label", f = /*wallet*/
      t[0].name + " download link"), p(e, "href", b = /*wallet*/
      t[0].download), p(e, "target", "_blank"), p(e, "rel", "noopener noreferrer"), p(e, "class", `rounded-md focus:outline-none  focus:ring-2 
    focus:ring-neutral-200  dark:focus:ring-neutral-700 transition-colors`);
    },
    m(c, v) {
      W(c, e, v), _(e, r), _(r, n), _(r, o), _(r, s), _(s, d), _(s, w), _(r, l), _(r, a), x || (m = [
        O(
          r,
          "click",
          /*click_handler*/
          t[6]
        ),
        O(
          r,
          "keyup",
          /*keyup_handler*/
          t[7]
        )
      ], x = !0);
    },
    p(c, v) {
      v & /*wallet*/
      1 && h !== (h = /*wallet*/
      c[0].name + "") && $(w, h), v & /*wallet*/
      1 && g !== (g = /*wallet*/
      c[0].name) && p(a, "alt", g), v & /*wallet*/
      1 && f !== (f = /*wallet*/
      c[0].name + " download link") && p(e, "aria-label", f), v & /*wallet*/
      1 && b !== (b = /*wallet*/
      c[0].download) && p(e, "href", b);
    },
    d(c) {
      c && E(e), x = !1, j(m);
    }
  };
}
function Ct(t) {
  let e, r, n;
  return {
    c() {
      e = B("img"), p(e, "alt", r = /*wallet*/
      t[0]?.name), _e(e.src, n = /*icon*/
      t[3]) || p(e, "src", n), p(e, "class", "w-8 h-8 rounded");
    },
    m(o, s) {
      W(o, e, s);
    },
    p(o, s) {
      s & /*wallet*/
      1 && r !== (r = /*wallet*/
      o[0]?.name) && p(e, "alt", r);
    },
    d(o) {
      o && E(e);
    }
  };
}
function At(t) {
  let e;
  return {
    c() {
      e = B("div"), Me(e, "position", "relative");
    },
    m(r, n) {
      W(r, e, n), e.innerHTML = /*icon*/
      t[3];
    },
    p: R,
    d(r) {
      r && E(e);
    }
  };
}
function Ot(t) {
  let e;
  return {
    c() {
      e = B("div"), e.innerHTML = '<svg aria-hidden="true" class="w-8 h-8 text-neutral-300 animate-spin dark:text-neutral-600 fill-neutral-600 dark:fill-neutral-300" viewBox="0 0 100 101" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M100 50.5908C100 78.2051 77.6142 100.591 50 100.591C22.3858 100.591 0 78.2051 0 50.5908C0 22.9766 22.3858 0.59082 50 0.59082C77.6142 0.59082 100 22.9766 100 50.5908ZM9.08144 50.5908C9.08144 73.1895 27.4013 91.5094 50 91.5094C72.5987 91.5094 90.9186 73.1895 90.9186 50.5908C90.9186 27.9921 72.5987 9.67226 50 9.67226C27.4013 9.67226 9.08144 27.9921 9.08144 50.5908Z" fill="currentColor"></path><path d="M93.9676 39.0409C96.393 38.4038 97.8624 35.9116 97.0079 33.5539C95.2932 28.8227 92.871 24.3692 89.8167 20.348C85.8452 15.1192 80.8826 10.7238 75.2124 7.41289C69.5422 4.10194 63.2754 1.94025 56.7698 1.05124C51.7666 0.367541 46.6976 0.446843 41.7345 1.27873C39.2613 1.69328 37.813 4.19778 38.4501 6.62326C39.0873 9.04874 41.5694 10.4717 44.0505 10.1071C47.8511 9.54855 51.7191 9.52689 55.5402 10.0491C60.8642 10.7766 65.9928 12.5457 70.6331 15.2552C75.2735 17.9648 79.3347 21.5619 82.5849 25.841C84.9175 28.9121 86.7997 32.2913 88.1811 35.8758C89.083 38.2158 91.5421 39.6781 93.9676 39.0409Z" fill="currentFill"></path></svg> <span class="sr-only">Loading...</span>', p(e, "role", "status");
    },
    m(r, n) {
      W(r, e, n);
    },
    p: R,
    d(r) {
      r && E(e);
    }
  };
}
function Nt(t) {
  let e;
  function r(s, d) {
    return (
      /*wallet*/
      s[0].download ? Ft : Bt
    );
  }
  let n = r(t), o = n(t);
  return {
    c() {
      o.c(), e = xe();
    },
    m(s, d) {
      o.m(s, d), W(s, e, d);
    },
    p(s, [d]) {
      n === (n = r(s)) && o ? o.p(s, d) : (o.d(1), o = n(s), o && (o.c(), o.m(e.parentNode, e)));
    },
    i: R,
    o: R,
    d(s) {
      s && E(e), o.d(s);
    }
  };
}
function Et(t, e, r) {
  let { wallet: n } = e, { theme: o = null } = e, { cb: s = async () => {
  } } = e, { loadingItem: d = !1 } = e;
  const h = typeof n.icon == "string" ? n.icon : o === "dark" ? n.icon.dark : n.icon.light, w = h?.startsWith("<svg"), l = () => {
    s(null);
  }, a = (f) => {
    f.key === "Enter" && s(null);
  }, g = async () => {
    s(n.connector);
  }, y = async (f) => {
    f.key === "Enter" && s(n.connector);
  };
  return t.$$set = (f) => {
    "wallet" in f && r(0, n = f.wallet), "theme" in f && r(5, o = f.theme), "cb" in f && r(1, s = f.cb), "loadingItem" in f && r(2, d = f.loadingItem);
  }, [
    n,
    s,
    d,
    h,
    w,
    o,
    l,
    a,
    g,
    y
  ];
}
class Pt extends Ne {
  constructor(e) {
    super(), Oe(this, e, Et, Nt, ke, {
      wallet: 0,
      theme: 5,
      cb: 1,
      loadingItem: 2
    });
  }
}
function we(t, e, r) {
  const n = t.slice();
  return n[16] = e[r], n;
}
function ge(t) {
  let e, r, n, o, s, d, h, w, l, a, g, y, f, b, x, m = fe(
    /*modalWallets*/
    t[1]
  ), c = [];
  for (let u = 0; u < m.length; u += 1)
    c[u] = he(we(t, m, u));
  const v = (u) => H(c[u], 1, 1, () => {
    c[u] = null;
  });
  return {
    c() {
      e = B("div"), r = B("main"), n = B("header"), o = B("h2"), o.textContent = "Connect to", s = z(), d = B("h1"), h = I(
        /*dappName*/
        t[0]
      ), w = z(), l = B("span"), l.innerHTML = '<svg width="12" height="12" viewBox="0 0 12 12" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M9.77275 3.02275C9.99242 2.80308 9.99242 2.44692 9.77275 2.22725C9.55308 2.00758 9.19692 2.00758 8.97725 2.22725L6 5.20451L3.02275 2.22725C2.80308 2.00758 2.44692 2.00758 2.22725 2.22725C2.00758 2.44692 2.00758 2.80308 2.22725 3.02275L5.20451 6L2.22725 8.97725C2.00758 9.19692 2.00758 9.55308 2.22725 9.77275C2.44692 9.99242 2.80308 9.99242 3.02275 9.77275L6 6.79549L8.97725 9.77275C9.19692 9.99242 9.55308 9.99242 9.77275 9.77275C9.99242 9.55308 9.99242 9.19692 9.77275 8.97725L6.79549 6L9.77275 3.02275Z" fill="currentColor"></path></svg>', a = z(), g = B("ul");
      for (let u = 0; u < c.length; u += 1)
        c[u].c();
      p(o, "class", "text-sm text-gray-400 font-semibold"), p(d, "class", `text-xl font-semibold mb-6 
                  max-w-[240px] overflow-hidden 
                  whitespace-nowrap text-ellipsis`), p(l, "class", `absolute top-0 right-0 p-2 cursor-pointer
                  rounded-full bg-neutral-100 dark:bg-neutral-800
                  text-neutral-400 dark:text-white
                  hover:bg-neutral-100 dark:hover:bg-neutral-700
                  focus:outline-none focus:ring-2
                focus:ring-neutral-200 dark:focus:ring-neutral-700
                  transition-colors`), p(l, "role", "button"), p(l, "tabindex", "0"), p(l, "aria-label", "Close"), p(n, "class", "flex items-center justify-center flex-col mb-2 relative"), p(g, "class", "flex flex-col gap-3"), p(r, "role", "dialog"), p(r, "class", `rounded-3xl shadow-modal dark:shadow-none 
              w-full max-w-[380px] z-50 
              mx-6 p-6 pb-8 text-center 
              bg-slate-50 dark:bg-neutral-900 
            text-neutral-900 dark:text-white`), p(e, "part", "starknetkit-modal"), p(e, "class", y = `modal-font backdrop-blur-sm fixed inset-0 flex items-center 
            justify-center bg-black/25 z-[9999] ${/*darkModeControlClass*/
      t[4]}`);
    },
    m(u, C) {
      W(u, e, C), _(e, r), _(r, n), _(n, o), _(n, s), _(n, d), _(d, h), _(n, w), _(n, l), _(r, a), _(r, g);
      for (let S = 0; S < c.length; S += 1)
        c[S] && c[S].m(g, null);
      f = !0, b || (x = [
        O(
          l,
          "click",
          /*click_handler*/
          t[8]
        ),
        O(
          l,
          "keyup",
          /*keyup_handler*/
          t[9]
        ),
        O(r, "click", Lt),
        O(r, "keyup", jt),
        O(
          e,
          "click",
          /*click_handler_2*/
          t[10]
        ),
        O(
          e,
          "keyup",
          /*keyup_handler_2*/
          t[11]
        )
      ], b = !0);
    },
    p(u, C) {
      if ((!f || C & /*dappName*/
      1) && $(
        h,
        /*dappName*/
        u[0]
      ), C & /*modalWallets, loadingItem, cb, theme*/
      78) {
        m = fe(
          /*modalWallets*/
          u[1]
        );
        let S;
        for (S = 0; S < m.length; S += 1) {
          const oe = we(u, m, S);
          c[S] ? (c[S].p(oe, C), L(c[S], 1)) : (c[S] = he(oe), c[S].c(), L(c[S], 1), c[S].m(g, null));
        }
        for (Be(), S = m.length; S < c.length; S += 1)
          v(S);
        Fe();
      }
      (!f || C & /*darkModeControlClass*/
      16 && y !== (y = `modal-font backdrop-blur-sm fixed inset-0 flex items-center 
            justify-center bg-black/25 z-[9999] ${/*darkModeControlClass*/
      u[4]}`)) && p(e, "class", y);
    },
    i(u) {
      if (!f) {
        for (let C = 0; C < m.length; C += 1)
          L(c[C]);
        f = !0;
      }
    },
    o(u) {
      c = c.filter(Boolean);
      for (let C = 0; C < c.length; C += 1)
        H(c[C]);
      f = !1;
    },
    d(u) {
      u && E(e), ht(c, u), b = !1, j(x);
    }
  };
}
function he(t) {
  let e, r;
  return e = new Pt({
    props: {
      wallet: (
        /*wallet*/
        t[16]
      ),
      loadingItem: (
        /*loadingItem*/
        t[3]
      ),
      cb: (
        /*cb*/
        t[6]
      ),
      theme: (
        /*theme*/
        t[2]
      )
    }
  }), {
    c() {
      xt(e.$$.fragment);
    },
    m(n, o) {
      Ce(e, n, o), r = !0;
    },
    p(n, o) {
      const s = {};
      o & /*modalWallets*/
      2 && (s.wallet = /*wallet*/
      n[16]), o & /*loadingItem*/
      8 && (s.loadingItem = /*loadingItem*/
      n[3]), o & /*theme*/
      4 && (s.theme = /*theme*/
      n[2]), e.$set(s);
    },
    i(n) {
      r || (L(e.$$.fragment, n), r = !0);
    },
    o(n) {
      H(e.$$.fragment, n), r = !1;
    },
    d(n) {
      Ae(e, n);
    }
  };
}
function zt(t) {
  let e, r, n = !/*isInAppBrowser*/
  t[5] && /*modalWallets*/
  t[1].length > 1 && ge(t);
  return {
    c() {
      n && n.c(), e = xe();
    },
    m(o, s) {
      n && n.m(o, s), W(o, e, s), r = !0;
    },
    p(o, [s]) {
      !/*isInAppBrowser*/
      o[5] && /*modalWallets*/
      o[1].length > 1 ? n ? (n.p(o, s), s & /*modalWallets*/
      2 && L(n, 1)) : (n = ge(o), n.c(), L(n, 1), n.m(e.parentNode, e)) : n && (Be(), H(n, 1, 1, () => {
        n = null;
      }), Fe());
    },
    i(o) {
      r || (L(n), r = !0);
    },
    o(o) {
      H(n), r = !1;
    },
    d(o) {
      o && E(e), n && n.d(o);
    }
  };
}
const Lt = (t) => t.stopPropagation(), jt = (t) => {
  t.stopPropagation();
};
function Wt(t, e, r) {
  let { dappName: n = window?.document.title ?? "" } = e, { modalWallets: o } = e, { callback: s = async () => {
  } } = e, { theme: d = null } = e, h = !1, l = window?.starknet_argentX?.isInAppBrowser;
  const g = navigator.userAgent.toLowerCase().includes("braavos"), y = (u) => {
    r(3, h = u);
  };
  let f = async (u) => {
    y(u?.id ?? !1);
    try {
      await s(u ?? null);
    } finally {
      y(!1);
    }
  }, b = d === "dark" ? "dark" : "";
  mt(async () => {
    if (d === "dark" || d === null && window.matchMedia("(prefers-color-scheme: dark)").matches ? r(4, b = "dark") : r(4, b = ""), l && window?.starknet_argentX) {
      try {
        s(new A({ options: { id: "argentX" } }));
      } catch {
      }
      return;
    }
    if (g && window?.starknet_braavos) {
      try {
        s(new A({ options: { id: "braavos" } }));
      } catch {
      }
      return;
    }
    if (o.length === 1)
      try {
        const [u] = o;
        await s(u.connector);
      } catch (u) {
        console.error(u);
      }
  });
  const x = () => f(null), m = (u) => {
    u.key === "Enter" && f(null);
  }, c = () => f(null), v = (u) => {
    u.key === "Escape" && f(null);
  };
  return t.$$set = (u) => {
    "dappName" in u && r(0, n = u.dappName), "modalWallets" in u && r(1, o = u.modalWallets), "callback" in u && r(7, s = u.callback), "theme" in u && r(2, d = u.theme);
  }, [
    n,
    o,
    d,
    h,
    b,
    l,
    f,
    s,
    x,
    m,
    c,
    v
  ];
}
class Tt extends Ne {
  constructor(e) {
    super(), Oe(this, e, Wt, zt, ke, {
      dappName: 0,
      modalWallets: 1,
      callback: 7,
      theme: 2
    });
  }
}
const Rt = `@import"https://fonts.googleapis.com/css2?family=Barlow:wght@500;600&display=swap";.sr-only{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);white-space:nowrap;border-width:0}.visible{visibility:visible}.static{position:static}.fixed{position:fixed}.absolute{position:absolute}.relative{position:relative}.inset-0{inset:0}.right-0{right:0}.top-0{top:0}.z-50{z-index:50}.z-\\[9999\\]{z-index:9999}.mx-6{margin-left:1.5rem;margin-right:1.5rem}.mb-2{margin-bottom:.5rem}.mb-6{margin-bottom:1.5rem}.block{display:block}.inline{display:inline}.flex{display:flex}.h-8{height:2rem}.w-8{width:2rem}.w-full{width:100%}.max-w-\\[240px\\]{max-width:240px}.max-w-\\[380px\\]{max-width:380px}.transform{transform:translate(var(--tw-translate-x),var(--tw-translate-y)) rotate(var(--tw-rotate)) skew(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y))}@keyframes spin{to{transform:rotate(360deg)}}.animate-spin{animation:spin 1s linear infinite}.cursor-pointer{cursor:pointer}.flex-row-reverse{flex-direction:row-reverse}.flex-col{flex-direction:column}.items-center{align-items:center}.justify-center{justify-content:center}.justify-between{justify-content:space-between}.gap-3{gap:.75rem}.overflow-hidden{overflow:hidden}.text-ellipsis{text-overflow:ellipsis}.whitespace-nowrap{white-space:nowrap}.rounded{border-radius:.25rem}.rounded-3xl{border-radius:1.5rem}.rounded-full{border-radius:9999px}.rounded-md{border-radius:.375rem}.border{border-width:1px}.bg-black\\/25{background-color:#00000040}.bg-neutral-100{--tw-bg-opacity: 1;background-color:rgb(245 245 245 / var(--tw-bg-opacity, 1))}.bg-slate-50{--tw-bg-opacity: 1;background-color:rgb(248 250 252 / var(--tw-bg-opacity, 1))}.fill-neutral-600{fill:#525252}.p-2{padding:.5rem}.p-3{padding:.75rem}.p-6{padding:1.5rem}.pb-8{padding-bottom:2rem}.text-center{text-align:center}.text-base{font-size:1rem;line-height:1.5rem}.text-sm{font-size:.875rem;line-height:1.25rem}.text-xl{font-size:1.25rem;line-height:1.75rem}.font-semibold{font-weight:600}.text-gray-400{--tw-text-opacity: 1;color:rgb(156 163 175 / var(--tw-text-opacity, 1))}.text-neutral-300{--tw-text-opacity: 1;color:rgb(212 212 212 / var(--tw-text-opacity, 1))}.text-neutral-400{--tw-text-opacity: 1;color:rgb(163 163 163 / var(--tw-text-opacity, 1))}.text-neutral-900{--tw-text-opacity: 1;color:rgb(23 23 23 / var(--tw-text-opacity, 1))}.shadow-list-item{--tw-shadow: 0px 2px 12px rgba(0, 0, 0, .12);--tw-shadow-colored: 0px 2px 12px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}.shadow-modal{--tw-shadow: 0px 4px 20px rgba(0, 0, 0, .5);--tw-shadow-colored: 0px 4px 20px var(--tw-shadow-color);box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}.outline{outline-style:solid}.blur{--tw-blur: blur(8px);filter:var(--tw-blur) var(--tw-brightness) var(--tw-contrast) var(--tw-grayscale) var(--tw-hue-rotate) var(--tw-invert) var(--tw-saturate) var(--tw-sepia) var(--tw-drop-shadow)}.filter{filter:var(--tw-blur) var(--tw-brightness) var(--tw-contrast) var(--tw-grayscale) var(--tw-hue-rotate) var(--tw-invert) var(--tw-saturate) var(--tw-sepia) var(--tw-drop-shadow)}.backdrop-blur-sm{--tw-backdrop-blur: blur(4px);-webkit-backdrop-filter:var(--tw-backdrop-blur) var(--tw-backdrop-brightness) var(--tw-backdrop-contrast) var(--tw-backdrop-grayscale) var(--tw-backdrop-hue-rotate) var(--tw-backdrop-invert) var(--tw-backdrop-opacity) var(--tw-backdrop-saturate) var(--tw-backdrop-sepia);backdrop-filter:var(--tw-backdrop-blur) var(--tw-backdrop-brightness) var(--tw-backdrop-contrast) var(--tw-backdrop-grayscale) var(--tw-backdrop-hue-rotate) var(--tw-backdrop-invert) var(--tw-backdrop-opacity) var(--tw-backdrop-saturate) var(--tw-backdrop-sepia)}.transition-colors{transition-property:color,background-color,border-color,text-decoration-color,fill,stroke;transition-timing-function:cubic-bezier(.4,0,.2,1);transition-duration:.15s}*,:before,:after{--tw-border-spacing-x: 0;--tw-border-spacing-y: 0;--tw-translate-x: 0;--tw-translate-y: 0;--tw-rotate: 0;--tw-skew-x: 0;--tw-skew-y: 0;--tw-scale-x: 1;--tw-scale-y: 1;--tw-pan-x: ;--tw-pan-y: ;--tw-pinch-zoom: ;--tw-scroll-snap-strictness: proximity;--tw-gradient-from-position: ;--tw-gradient-via-position: ;--tw-gradient-to-position: ;--tw-ordinal: ;--tw-slashed-zero: ;--tw-numeric-figure: ;--tw-numeric-spacing: ;--tw-numeric-fraction: ;--tw-ring-inset: ;--tw-ring-offset-width: 0px;--tw-ring-offset-color: #fff;--tw-ring-color: rgb(59 130 246 / .5);--tw-ring-offset-shadow: 0 0 #0000;--tw-ring-shadow: 0 0 #0000;--tw-shadow: 0 0 #0000;--tw-shadow-colored: 0 0 #0000;--tw-blur: ;--tw-brightness: ;--tw-contrast: ;--tw-grayscale: ;--tw-hue-rotate: ;--tw-invert: ;--tw-saturate: ;--tw-sepia: ;--tw-drop-shadow: ;--tw-backdrop-blur: ;--tw-backdrop-brightness: ;--tw-backdrop-contrast: ;--tw-backdrop-grayscale: ;--tw-backdrop-hue-rotate: ;--tw-backdrop-invert: ;--tw-backdrop-opacity: ;--tw-backdrop-saturate: ;--tw-backdrop-sepia: ;--tw-contain-size: ;--tw-contain-layout: ;--tw-contain-paint: ;--tw-contain-style: }::backdrop{--tw-border-spacing-x: 0;--tw-border-spacing-y: 0;--tw-translate-x: 0;--tw-translate-y: 0;--tw-rotate: 0;--tw-skew-x: 0;--tw-skew-y: 0;--tw-scale-x: 1;--tw-scale-y: 1;--tw-pan-x: ;--tw-pan-y: ;--tw-pinch-zoom: ;--tw-scroll-snap-strictness: proximity;--tw-gradient-from-position: ;--tw-gradient-via-position: ;--tw-gradient-to-position: ;--tw-ordinal: ;--tw-slashed-zero: ;--tw-numeric-figure: ;--tw-numeric-spacing: ;--tw-numeric-fraction: ;--tw-ring-inset: ;--tw-ring-offset-width: 0px;--tw-ring-offset-color: #fff;--tw-ring-color: rgb(59 130 246 / .5);--tw-ring-offset-shadow: 0 0 #0000;--tw-ring-shadow: 0 0 #0000;--tw-shadow: 0 0 #0000;--tw-shadow-colored: 0 0 #0000;--tw-blur: ;--tw-brightness: ;--tw-contrast: ;--tw-grayscale: ;--tw-hue-rotate: ;--tw-invert: ;--tw-saturate: ;--tw-sepia: ;--tw-drop-shadow: ;--tw-backdrop-blur: ;--tw-backdrop-brightness: ;--tw-backdrop-contrast: ;--tw-backdrop-grayscale: ;--tw-backdrop-hue-rotate: ;--tw-backdrop-invert: ;--tw-backdrop-opacity: ;--tw-backdrop-saturate: ;--tw-backdrop-sepia: ;--tw-contain-size: ;--tw-contain-layout: ;--tw-contain-paint: ;--tw-contain-style: }*,:before,:after{box-sizing:border-box;border-width:0;border-style:solid;border-color:#e5e7eb}:before,:after{--tw-content: ""}html,:host{line-height:1.5;-webkit-text-size-adjust:100%;-moz-tab-size:4;-o-tab-size:4;tab-size:4;font-family:ui-sans-serif,system-ui,sans-serif,"Apple Color Emoji","Segoe UI Emoji",Segoe UI Symbol,"Noto Color Emoji";font-feature-settings:normal;font-variation-settings:normal;-webkit-tap-highlight-color:transparent}body{margin:0;line-height:inherit}hr{height:0;color:inherit;border-top-width:1px}abbr:where([title]){-webkit-text-decoration:underline dotted;text-decoration:underline dotted}h1,h2,h3,h4,h5,h6{font-size:inherit;font-weight:inherit}a{color:inherit;text-decoration:inherit}b,strong{font-weight:bolder}code,kbd,samp,pre{font-family:ui-monospace,SFMono-Regular,Menlo,Monaco,Consolas,Liberation Mono,Courier New,monospace;font-feature-settings:normal;font-variation-settings:normal;font-size:1em}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}table{text-indent:0;border-color:inherit;border-collapse:collapse}button,input,optgroup,select,textarea{font-family:inherit;font-feature-settings:inherit;font-variation-settings:inherit;font-size:100%;font-weight:inherit;line-height:inherit;letter-spacing:inherit;color:inherit;margin:0;padding:0}button,select{text-transform:none}button,input:where([type=button]),input:where([type=reset]),input:where([type=submit]){-webkit-appearance:button;background-color:transparent;background-image:none}:-moz-focusring{outline:auto}:-moz-ui-invalid{box-shadow:none}progress{vertical-align:baseline}::-webkit-inner-spin-button,::-webkit-outer-spin-button{height:auto}[type=search]{-webkit-appearance:textfield;outline-offset:-2px}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}summary{display:list-item}blockquote,dl,dd,h1,h2,h3,h4,h5,h6,hr,figure,p,pre{margin:0}fieldset{margin:0;padding:0}legend{padding:0}ol,ul,menu{list-style:none;margin:0;padding:0}dialog{padding:0}textarea{resize:vertical}input::-moz-placeholder,textarea::-moz-placeholder{opacity:1;color:#9ca3af}input::placeholder,textarea::placeholder{opacity:1;color:#9ca3af}button,[role=button]{cursor:pointer}:disabled{cursor:default}img,svg,video,canvas,audio,iframe,embed,object{display:block;vertical-align:middle}img,video{max-width:100%;height:auto}[hidden]:where(:not([hidden=until-found])){display:none}.modal-font{font-family:Barlow,-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Oxygen,Ubuntu,Cantarell,Fira Sans,Droid Sans,Helvetica Neue,sans-serif;-webkit-font-smoothing:antialiased;text-rendering:optimizeLegibility;-webkit-text-size-adjust:100%;-moz-text-size-adjust:100%;text-size-adjust:100%;font-feature-settings:"kern"}.l2{color:#8c8c8c;font-size:12px;font-weight:500;line-height:14px;letter-spacing:0em;text-align:left}.p{margin:0}.hover\\:bg-neutral-100:hover{--tw-bg-opacity: 1;background-color:rgb(245 245 245 / var(--tw-bg-opacity, 1))}.focus\\:outline-none:focus{outline:2px solid transparent;outline-offset:2px}.focus\\:ring-2:focus{--tw-ring-offset-shadow: var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color);--tw-ring-shadow: var(--tw-ring-inset) 0 0 0 calc(2px + var(--tw-ring-offset-width)) var(--tw-ring-color);box-shadow:var(--tw-ring-offset-shadow),var(--tw-ring-shadow),var(--tw-shadow, 0 0 #0000)}.focus\\:ring-neutral-200:focus{--tw-ring-opacity: 1;--tw-ring-color: rgb(229 229 229 / var(--tw-ring-opacity, 1))}.dark\\:bg-neutral-800:is(.dark *){--tw-bg-opacity: 1;background-color:rgb(38 38 38 / var(--tw-bg-opacity, 1))}.dark\\:bg-neutral-900:is(.dark *){--tw-bg-opacity: 1;background-color:rgb(23 23 23 / var(--tw-bg-opacity, 1))}.dark\\:fill-neutral-300:is(.dark *){fill:#d4d4d4}.dark\\:text-neutral-600:is(.dark *){--tw-text-opacity: 1;color:rgb(82 82 82 / var(--tw-text-opacity, 1))}.dark\\:text-white:is(.dark *){--tw-text-opacity: 1;color:rgb(255 255 255 / var(--tw-text-opacity, 1))}.dark\\:shadow-none:is(.dark *){--tw-shadow: 0 0 #0000;--tw-shadow-colored: 0 0 #0000;box-shadow:var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow)}.dark\\:hover\\:bg-neutral-700:hover:is(.dark *){--tw-bg-opacity: 1;background-color:rgb(64 64 64 / var(--tw-bg-opacity, 1))}.dark\\:focus\\:ring-neutral-700:focus:is(.dark *){--tw-ring-opacity: 1;--tw-ring-color: rgb(64 64 64 / var(--tw-ring-opacity, 1))}
`, Kt = (t) => ({
  starknetkitConnectModal: async () => await It({
    ...t,
    resultType: t?.resultType ?? "connector"
  })
});
let N = null;
const It = async ({
  modalMode: t = "canAsk",
  storeVersion: e = He(),
  modalTheme: r,
  dappName: n,
  resultType: o = "wallet",
  ...s
}) => {
  const { webWalletUrl: d = je, argentMobileOptions: h } = s, { connectors: w } = s;
  N = null;
  const l = !w || w.length === 0 ? Ie({
    argentMobileOptions: h,
    webWalletUrl: d
  }) : w, a = localStorage.getItem("starknetLastConnectedWallet");
  if (t === "neverAsk")
    try {
      const b = l.find((m) => m.id === a) ?? null;
      let x = null;
      return b && o === "wallet" && (x = await b.connect()), {
        connector: b,
        wallet: b?.wallet ?? null,
        connectorData: x
      };
    } catch (b) {
      throw pe(), new Error(b);
    }
  const g = await Z.getAvailableWallets(s);
  if (t === "canAsk" && a && ((await Z.getAuthorizedWallets(s)).find((m) => m.id === a) ?? g.length === 1 ? g[0] : void 0)) {
    const m = l.find((v) => v.id === a);
    let c = null;
    return o === "wallet" && (c = await m?.connect() ?? null), m && (N = m), {
      connector: N,
      connectorData: c,
      wallet: m?.wallet ?? null
    };
  }
  const y = wt({
    availableConnectors: l,
    installedWallets: g,
    discoveryWallets: await Z.getDiscoveryWallets(s),
    storeVersion: e,
    customOrder: w ? w?.length > 0 : !1
  }), f = () => {
    const b = "starknetkit-modal-container", x = document.getElementById(b);
    if (x) {
      if (x.shadowRoot)
        return x.shadowRoot;
      x.remove();
    }
    const m = document.createElement("div");
    m.id = b, document.body.appendChild(m);
    const c = m.attachShadow({ mode: "open" });
    return c.innerHTML = `<style>${Rt}</style>`, c;
  };
  return new Promise((b, x) => {
    const m = new Tt({
      target: f(),
      props: {
        dappName: n,
        callback: async (c) => {
          try {
            if (N = c, o === "wallet") {
              const v = await c?.connect() ?? null;
              c !== null && ze(c.id), b({
                connector: c,
                connectorData: v,
                wallet: c?.wallet ?? null
              });
            } else
              b({
                connector: c,
                wallet: null,
                connectorData: null
              });
          } catch (v) {
            x(v);
          } finally {
            setTimeout(() => m.$destroy());
          }
        },
        theme: r === "system" ? null : r ?? null,
        modalWallets: y
      }
    });
  });
}, Yt = () => N ? N.wallet : null, $t = async (t = {}) => (pe(), N && await N.disconnect(), N = null, Z.disconnect(t));
export {
  It as connect,
  $t as disconnect,
  Yt as getSelectedConnectorWallet,
  Kt as useStarknetkitConnectModal
};
